// @nitrostack/core/dist/core/app-decorator.js
import 'reflect-metadata';
import { createServer } from './server.js';
import { createLogger } from './logger.js';
import { DIContainer } from './di/container.js';
import { registerHealthCheck } from './decorators/health-check.decorator.js';
import { registerEventHandlers } from './events/event.decorator.js';
import { createResource } from './resource.js';
import { getModuleMetadata } from './module.js';
import { buildTools, buildResources, buildPrompts } from './builders.js';
/**
 * Helper to check if a provider is a ClassConstructor
 */
function isClassConstructor(provider) {
    return typeof provider === 'function';
}
/**
 * Register a provider (either class or object-style provider)
 */
function registerProvider(container, provider) {
    if (isClassConstructor(provider)) {
        container.register(provider);
    }
    else {
        // Object-style provider: { provide: ..., useValue/useClass/useFactory: ... }
        const token = provider.provide;
        if (provider.useValue !== undefined) {
            container.registerValue(token, provider.useValue);
        }
        else if (provider.useClass) {
            container.register(token, provider.useClass);
        }
        // useFactory could be added later if needed
    }
}
/**
 * MCP Application Metadata Key
 */
const MCP_APP_METADATA = Symbol('mcp:app');
/**
 * @McpApp Decorator
 * Marks a class as the root application module
 */
export function McpApp(options) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (target) => {
        Reflect.defineMetadata(MCP_APP_METADATA, options, target);
        return target;
    };
}
/**
 * Get MCP App metadata
 */
export function getMcpAppMetadata(target) {
    return Reflect.getMetadata(MCP_APP_METADATA, target);
}
/**
 * MCP Application Factory
 * Creates and bootstraps an MCP application
 */
export class McpApplicationFactory {
    /**
     * Create an MCP application from an app module
     */
    static async create(AppModule) {
        const options = getMcpAppMetadata(AppModule);
        if (!options) {
            throw new Error(`${AppModule.name} is not decorated with @McpApp. ` +
                `Please add @McpApp decorator to your root module.`);
        }
        // Create logger
        const logger = createLogger({ level: options.logging?.level || 'info' });
        // Set log level if provided
        if (options.logging?.level) {
            process.env.LOG_LEVEL = options.logging.level;
        }
        // Create DI container
        const container = DIContainer.getInstance();
        // Register logger in DI for modules that need it
        container.registerValue('Logger', logger);
        // Get module metadata
        const moduleMetadata = getModuleMetadata(options.module);
        if (!moduleMetadata) {
            throw new Error(`${options.module.name} is not decorated with @Module`);
        }
        // Register all providers
        if (moduleMetadata.providers) {
            for (const provider of moduleMetadata.providers) {
                registerProvider(container, provider);
                // Check if provider is a health check or event handler (only for class constructors)
                if (isClassConstructor(provider)) {
                    const healthCheckMetadata = Reflect.getMetadata('nitrostack:health_check', provider);
                    if (healthCheckMetadata) {
                        // Resolve and register health check
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        const healthCheckInstance = container.resolve(provider);
                        registerHealthCheck(healthCheckInstance, healthCheckMetadata);
                        logger.info(`  ✓ Health Check: ${healthCheckMetadata.name}`);
                    }
                    // Check for event handlers
                    const eventHandlers = Reflect.getMetadata('nitrostack:event_handler', provider);
                    if (eventHandlers && eventHandlers.length > 0) {
                        // Resolve and register event handlers
                        const instance = container.resolve(provider);
                        registerEventHandlers(instance);
                        logger.info(`  ✓ Event Handler: ${provider.name} (${eventHandlers.length} event(s))`);
                    }
                }
            }
        }
        // Register all controllers (tools/resources/prompts)
        if (moduleMetadata.controllers) {
            for (const controller of moduleMetadata.controllers) {
                container.register(controller);
            }
        }
        // Track dynamic modules (forRoot() style) to add to server after creation
        const dynamicModulesToAdd = [];
        // Process imports (other modules)
        if (moduleMetadata.imports) {
            for (const importedModule of moduleMetadata.imports) {
                if (typeof importedModule === 'function') {
                    const importedMetadata = getModuleMetadata(importedModule);
                    if (importedMetadata) {
                        // Register providers from imported module
                        if (importedMetadata.providers) {
                            for (const provider of importedMetadata.providers) {
                                registerProvider(container, provider);
                                // Check if provider is a health check or event handler (only for class constructors)
                                if (isClassConstructor(provider)) {
                                    const healthCheckMetadata = Reflect.getMetadata('nitrostack:health_check', provider);
                                    if (healthCheckMetadata) {
                                        // Resolve and register health check
                                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                                        const healthCheckInstance = container.resolve(provider);
                                        registerHealthCheck(healthCheckInstance, healthCheckMetadata);
                                        logger.info(`  ✓ Health Check: ${healthCheckMetadata.name}`);
                                    }
                                    // Check for event handlers
                                    const eventHandlers = Reflect.getMetadata('nitrostack:event_handler', provider);
                                    if (eventHandlers && eventHandlers.length > 0) {
                                        // Resolve and register event handlers
                                        const instance = container.resolve(provider);
                                        registerEventHandlers(instance);
                                        logger.info(`  ✓ Event Handler: ${provider.name} (${eventHandlers.length} event(s))`);
                                    }
                                }
                            }
                        }
                        // Register controllers from imported module
                        if (importedMetadata.controllers) {
                            for (const controller of importedMetadata.controllers) {
                                container.register(controller);
                            }
                        }
                    }
                }
                else if (importedModule && typeof importedModule === 'object') {
                    // Handle ConfigModule.forRoot() style dynamic modules
                    const dynamicMod = importedModule;
                    if (dynamicMod.providers) {
                        for (const provider of dynamicMod.providers) {
                            // Check if provider has {provide, useValue} format
                            if (provider && typeof provider === 'object' && 'provide' in provider && 'useValue' in provider) {
                                const valueProvider = provider;
                                container.registerValue(valueProvider.provide, valueProvider.useValue);
                            }
                            else {
                                container.register(provider);
                            }
                        }
                    }
                    // If the dynamic module has a module class, we'll add it to the server after creation
                    // Don't register it in DI yet to avoid circular dependency issues
                    if (dynamicMod.module) {
                        dynamicModulesToAdd.push(dynamicMod.module);
                    }
                }
            }
        }
        // Create server
        const server = createServer({
            name: options.server?.name || 'mcp-server',
            version: options.server?.version || '1.0.0',
        });
        // Now register and add dynamic modules (from forRoot() calls) to server
        for (const dynamicModule of dynamicModulesToAdd) {
            container.register(dynamicModule);
            // Access internal modules array
            const serverInternal = server;
            serverInternal.modules.push(dynamicModule);
        }
        // Build and register tools, resources, and prompts from all controllers
        let totalTools = 0;
        let totalResources = 0;
        let totalPrompts = 0;
        // Process all registered controllers
        const allControllers = [];
        // Get controllers from root module
        if (moduleMetadata.controllers) {
            allControllers.push(...moduleMetadata.controllers);
        }
        // Get controllers from imported modules
        if (moduleMetadata.imports) {
            for (const importedModule of moduleMetadata.imports) {
                if (typeof importedModule === 'function') {
                    const importedMetadata = getModuleMetadata(importedModule);
                    if (importedMetadata?.controllers) {
                        allControllers.push(...importedMetadata.controllers);
                    }
                }
            }
        }
        // Build and register from each controller
        for (const controllerClass of allControllers) {
            // Resolve controller instance from DI container
            const controller = container.resolve(controllerClass);
            // Build tools
            const tools = buildTools(controller);
            for (const tool of tools) {
                server.tool(tool);
                totalTools++;
            }
            // Build resources
            const resources = buildResources(controller);
            for (const resource of resources) {
                server.resource(resource);
                totalResources++;
            }
            // Build prompts
            const prompts = buildPrompts(controller);
            for (const prompt of prompts) {
                server.prompt(prompt);
                totalPrompts++;
            }
        }
        // Register health checks resource if any health checks exist
        const { buildHealthChecksResource } = await import('./health/health-checks.resource.js');
        const healthChecksResourceDef = await buildHealthChecksResource();
        const healthChecksResource = createResource({
            uri: healthChecksResourceDef.uri,
            name: healthChecksResourceDef.name,
            description: healthChecksResourceDef.description,
            mimeType: healthChecksResourceDef.mimeType,
            handler: async (uri, context) => {
                const content = await healthChecksResourceDef.read();
                return { type: 'text', data: content };
            },
        });
        server.resource(healthChecksResource);
        logger.info(`✅ Health checks resource registered`);
        // Load widget manifest and register widget examples resource
        try {
            const { getWidgetRegistry } = await import('./widgets/widget-registry.js');
            const { buildWidgetExamplesResource } = await import('./widgets/widget-examples.resource.js');
            const { join } = await import('path');
            const registry = getWidgetRegistry();
            const widgetManifestPath = join(process.cwd(), 'src', 'widgets', 'widget-manifest.json');
            registry.loadManifest(widgetManifestPath);
            const widgetExamplesResourceDef = await buildWidgetExamplesResource();
            const widgetExamplesResource = createResource({
                uri: widgetExamplesResourceDef.uri,
                name: widgetExamplesResourceDef.name,
                description: widgetExamplesResourceDef.description,
                mimeType: widgetExamplesResourceDef.mimeType,
                handler: async (uri, context) => {
                    const content = await widgetExamplesResourceDef.read();
                    return { type: 'text', data: content };
                },
            });
            server.resource(widgetExamplesResource);
            logger.info(`Widget examples resource registered`);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            logger.warn(`Widget examples not available: ${errorMessage}`);
        }
        logger.info(`✅ Application initialized with ${totalTools} tools, ${totalResources} resources, ${totalPrompts} prompts`);
        // Auto-detect transport type based on configuration
        let transportType = 'stdio';
        let transportOptions = undefined;
        // Check explicit transport configuration
        if (options.transport?.type) {
            transportType = options.transport.type;
        }
        // Check if OAuth is configured
        const { OAuthModule } = await import('./oauth-module.js');
        const oauthConfig = OAuthModule.getConfig();
        if (oauthConfig) {
            // OAuth requires DUAL transport: STDIO for MCP + HTTP for metadata
            // This allows Studio to connect via STDIO while exposing OAuth metadata via HTTP
            transportType = 'dual';
            // Extract port from resourceUri (e.g., http://localhost:3002)
            let port = 3000;
            try {
                const resourceUrl = new URL(oauthConfig.resourceUri);
                port = resourceUrl.port ? parseInt(resourceUrl.port) : (resourceUrl.protocol === 'https:' ? 443 : 80);
            }
            catch (error) {
                logger.warn(`Failed to parse resourceUri for port, using default 3000`);
            }
            // Override with explicit config if provided
            if (oauthConfig.http?.port) {
                port = oauthConfig.http.port;
            }
            else if (options.transport?.http?.port) {
                port = options.transport.http.port;
            }
            transportOptions = {
                port,
                host: oauthConfig.http?.host || options.transport?.http?.host || '0.0.0.0',
                basePath: oauthConfig.http?.basePath || options.transport?.http?.basePath || '/mcp',
                oauth: {
                    resourceUri: oauthConfig.resourceUri,
                    authorizationServers: oauthConfig.authorizationServers,
                    scopesSupported: oauthConfig.scopesSupported,
                },
            };
            logger.info(`🔐 OAuth 2.1 detected - using HTTP transport on port ${port}`);
        }
        else if (options.transport?.type === 'http') {
            // Explicitly configured HTTP transport
            transportType = 'http';
            transportOptions = {
                port: options.transport.http?.port || 3000,
                host: options.transport.http?.host || '0.0.0.0',
                basePath: options.transport.http?.basePath || '/mcp',
            };
        }
        // Store transport configuration on server for later use
        const serverInternal = server;
        serverInternal._transportType = transportType;
        serverInternal._transportOptions = transportOptions;
        // Register graceful-shutdown signal handlers so shutdown lifecycle hooks run
        // on SIGTERM/SIGINT. Opt-out via `shutdownHooks: false`.
        if (options.shutdownHooks !== false) {
            server.enableShutdownHooks();
        }
        return server;
    }
}
