// @nitrostack/core/dist/core/builders.js
import { Tool } from './tool.js';
import { createResource } from './resource.js';
import { createPrompt } from './prompt.js';
import { createComponentFromNextRoute } from '../ui-next/index.js';
import { extractTools, extractResources, extractPrompts, getWidgetMetadata, getInitialToolMetadata, getControllerPrefix, } from './decorators.js';
import { getGuardsMetadata } from './guards/use-guards.decorator.js';
import { getMiddlewareMetadata } from './middleware/middleware.decorator.js';
import { getInterceptorMetadata } from './interceptors/interceptor.decorator.js';
import { getPipeMetadata } from './pipes/pipe.decorator.js';
import { getExceptionFilterMetadata } from './filters/exception-filter.decorator.js';
import { DIContainer } from './di/container.js';
/**
 * Build a Tool instance from a decorated method
 */
export function buildTool(controllerInstance, methodName, options, widgetMeta, isInitial) {
    const widgetRoute = widgetMeta?.route;
    const prototype = Object.getPrototypeOf(controllerInstance);
    // Get the original method
    const method = controllerInstance[methodName];
    const originalMethod = method.bind(controllerInstance);
    // Extract all decorator metadata
    const guards = getGuardsMetadata(prototype, methodName);
    const middlewares = getMiddlewareMetadata(prototype, methodName);
    const interceptors = getInterceptorMetadata(prototype, methodName);
    const pipes = getPipeMetadata(prototype, methodName);
    const filters = getExceptionFilterMetadata(prototype, methodName);
    // Create tool with all metadata
    const tool = new Tool({
        name: options.name,
        title: options.title,
        description: options.description,
        inputSchema: options.inputSchema,
        outputSchema: options.outputSchema,
        annotations: options.annotations,
        invocation: options.invocation,
        handler: originalMethod,
        guards,
        middlewares,
        interceptors,
        pipes,
        filters,
        examples: options.examples,
        widget: widgetRoute ? { route: widgetRoute } : undefined,
        outputTemplate: widgetRoute,
        isInitial,
        taskSupport: options.taskSupport,
    });
    // Create and attach component if widget route is specified
    if (widgetRoute) {
        const component = createComponentFromNextRoute(widgetRoute, {
            csp: widgetMeta?.csp,
            prefersBorder: widgetMeta?.prefersBorder,
            subdomain: widgetMeta?.domain,
        });
        tool.setComponent(component);
    }
    return tool;
}
/**
 * Build all tools from a controller
 */
export function buildTools(controllerInstance) {
    const constructor = Object.getPrototypeOf(controllerInstance).constructor;
    const toolMetadata = extractTools(constructor);
    const prefix = getControllerPrefix(constructor);
    return toolMetadata.map(({ methodName, options }) => {
        const prototype = Object.getPrototypeOf(controllerInstance);
        // Check if this method has a widget
        const widgetMeta = getWidgetMetadata(prototype, methodName);
        // Check for initial tool decorator
        const isInitial = getInitialToolMetadata(prototype, methodName);
        // Apply the controller prefix without mutating the shared options object
        // stored in class metadata. Skip when the name is already prefixed.
        const name = prefix && !options.name.startsWith(`${prefix}_`)
            ? `${prefix}_${options.name}`
            : options.name;
        const toolOptions = name === options.name ? options : { ...options, name };
        return buildTool(controllerInstance, methodName, toolOptions, widgetMeta, isInitial);
    });
}
/**
 * Build a Resource instance from a decorated method
 */
export function buildResource(controllerInstance, methodName, options) {
    const method = controllerInstance[methodName];
    const originalMethod = method.bind(controllerInstance);
    const resourceDefinition = {
        uri: options.uri,
        name: options.name,
        title: options.title,
        description: options.description,
        mimeType: options.mimeType,
        size: options.size,
        annotations: options.annotations,
        handler: async (uri, context) => {
            const result = await originalMethod(uri, context);
            // If result is already in ResourceContent format, return it
            if (result && typeof result === 'object' && ('type' in result) && ('data' in result)) {
                return result;
            }
            // Otherwise, infer type (default to JSON for objects/arrays, text for strings)
            if (typeof result === 'string') {
                return { type: 'text', data: result };
            }
            else if (Buffer.isBuffer(result)) {
                return { type: 'binary', data: result };
            }
            else {
                return { type: 'json', data: result };
            }
        },
        metadata: options.metadata,
    };
    return createResource(resourceDefinition);
}
/**
 * Build all resources from a controller
 */
export function buildResources(controllerInstance) {
    const constructor = Object.getPrototypeOf(controllerInstance).constructor;
    const resourceMetadata = extractResources(constructor);
    return resourceMetadata.map(({ methodName, options }) => {
        return buildResource(controllerInstance, methodName, options);
    });
}
/**
 * Build a Prompt instance from a decorated method
 */
export function buildPrompt(controllerInstance, methodName, options) {
    const method = controllerInstance[methodName];
    const originalMethod = method.bind(controllerInstance);
    const promptDefinition = {
        name: options.name,
        title: options.title,
        description: options.description,
        arguments: options.arguments,
        handler: async (args, context) => {
            return await originalMethod(args, context);
        },
    };
    return createPrompt(promptDefinition);
}
/**
 * Build all prompts from a controller
 */
export function buildPrompts(controllerInstance) {
    const constructor = Object.getPrototypeOf(controllerInstance).constructor;
    const promptMetadata = extractPrompts(constructor);
    return promptMetadata.map(({ methodName, options }) => {
        return buildPrompt(controllerInstance, methodName, options);
    });
}
/**
 * Build all tools, resources, and prompts from a controller
 * Supports dependency injection
 */
export function buildController(controller) {
    const container = DIContainer.getInstance();
    // Create controller instance (with DI if available)
    // Resolve from DI whenever the controller is registered so tools bind to the
    // same singleton that receives lifecycle hooks (via instantiateAll). Falling
    // back to `new controller()` only for unregistered controllers avoids the
    // double-instance divergence for non-@Injectable controllers.
    let instance;
    if (container.has(controller)) {
        instance = container.resolve(controller);
    }
    else {
        instance = new controller();
    }
    return {
        tools: buildTools(instance),
        resources: buildResources(instance),
        prompts: buildPrompts(instance),
    };
}
