// @nitrostack/core/dist/core/config-module.js
import 'reflect-metadata';
import * as fs from 'fs';
import * as path from 'path';
/**
 * Parse .env file manually to avoid stdout pollution
 * (dotenv.config() can write to stdout which breaks MCP stdio protocol)
 */
function parseEnvFile(filePath) {
    const envVars = {};
    if (!fs.existsSync(filePath)) {
        return envVars;
    }
    const content = fs.readFileSync(filePath, 'utf-8');
    content.split('\n').forEach(line => {
        line = line.trim();
        // Skip empty lines and comments
        if (!line || line.startsWith('#')) {
            return;
        }
        // Parse KEY=VALUE
        const [key, ...valueParts] = line.split('=');
        if (key && valueParts.length > 0) {
            let value = valueParts.join('=').trim();
            // Remove quotes if present
            if ((value.startsWith('"') && value.endsWith('"')) ||
                (value.startsWith("'") && value.endsWith("'"))) {
                value = value.slice(1, -1);
            }
            envVars[key.trim()] = value;
        }
    });
    return envVars;
}
/**
 * Configuration Service
 * Provides type-safe access to environment variables
 */
export class ConfigService {
    config;
    constructor(options = {}) {
        this.config = { ...options.defaults };
        // Load environment variables from .env file
        if (!options.ignoreEnvFile) {
            const envPath = options.envFilePath || '.env';
            const fullPath = path.resolve(process.cwd(), envPath);
            const parsed = parseEnvFile(fullPath);
            this.config = { ...this.config, ...parsed };
        }
        // Merge with process.env (already loaded by Studio)
        this.config = { ...this.config, ...process.env };
        // Validate if validator provided
        if (options.validate && !options.validate(this.config)) {
            throw new Error('Environment variable validation failed');
        }
    }
    get(key, defaultValue) {
        const value = this.config[key];
        if (value === undefined) {
            return defaultValue;
        }
        return value;
    }
    /**
     * Get a required configuration value
     * Throws if not found
     */
    getOrThrow(key) {
        const value = this.get(key);
        if (value === undefined) {
            throw new Error(`Configuration key "${key}" is required but not found`);
        }
        return value;
    }
    /**
     * Get all configuration
     */
    getAll() {
        return { ...this.config };
    }
}
/**
 * Configuration Module
 * NestJS-style configuration management
 */
export class ConfigModule {
    static forRoot(options = {}) {
        const configService = new ConfigService(options);
        return {
            providers: [{ provide: ConfigService, useValue: configService }],
            exports: [ConfigService],
        };
    }
}
