// @nitrostack/core/dist/core/decorators.js
import 'reflect-metadata';
import { DIContainer } from './di/container.js';
/**
 * Metadata keys for decorators
 */
export const TOOL_METADATA = Symbol('tool:metadata');
export const WIDGET_METADATA = Symbol('widget:metadata');
export const RESOURCE_METADATA = Symbol('resource:metadata');
export const PROMPT_METADATA = Symbol('prompt:metadata');
export const GUARDS_METADATA = Symbol('guards:metadata');
export const CONTROLLER_METADATA = Symbol('controller:metadata');
/**
 * Tool decorator - Marks a method as an MCP tool
 *
 * @example
 * ```typescript
 * @Tool({
 *   name: 'login',
 *   description: 'Login with email and password',
 *   inputSchema: z.object({
 *     email: z.string().email(),
 *     password: z.string(),
 *   }),
 * })
 * async login(input: LoginInput, context: ExecutionContext) {
 *   // Implementation
 * }
 * ```
 */
export function Tool(options) {
    return function (target, propertyKey, descriptor) {
        // Get existing tools or create new array. Clone to avoid mutating a parent
        // class's metadata array when subclassing (Reflect.getMetadata walks the
        // prototype chain, so a shared reference would pollute the parent).
        const existingTools = [...(Reflect.getMetadata(TOOL_METADATA, target.constructor) || [])];
        // Add this tool
        existingTools.push({
            methodName: String(propertyKey),
            options,
        });
        // Store metadata on the class constructor
        Reflect.defineMetadata(TOOL_METADATA, existingTools, target.constructor);
        return descriptor;
    };
}
/**
 * Widget decorator - Links a tool to a Next.js widget route
 *
 * @example String route (backward compatible)
 * ```typescript
 * @Widget('login-result')
 * ```
 *
 * @example Object form — `route` is required
 * ```typescript
 * @Widget({
 *   route: 'card',
 *   prefersBorder: true,
 *   domain: 'https://myapp.example.com',
 *   csp: {
 *     resourceDomains: ['https://images.unsplash.com'],
 *     connectDomains: ['https://api.example.com'],
 *     frameDomains: ['https://*.example-embed.com'],
 *   },
 * })
 * ```
 */
export function Widget(routePath) {
    let meta;
    if (typeof routePath === 'string') {
        if (!routePath.trim()) {
            throw new Error('@Widget: route must be a non-empty string');
        }
        meta = { route: routePath };
    }
    else {
        if (typeof routePath.route !== 'string' || !routePath.route.trim()) {
            throw new Error('@Widget: object form requires a non-empty string "route"');
        }
        meta = routePath;
    }
    return function (target, propertyKey, descriptor) {
        Reflect.defineMetadata(WIDGET_METADATA, meta, target, String(propertyKey));
        return descriptor;
    };
}
/**
 * Controller decorator - Marks a class as a controller, auto-registers it in
 * the DI container, and optionally prefixes all of its @Tool names.
 *
 * @example
 * ```typescript
 * @Controller('github')
 * export class GitHubController {
 *   @Tool({ name: 'create_issue', ... })
 *   async createIssue() { } // exposed as `github_create_issue`
 * }
 * ```
 */
export function Controller(prefixOrOptions) {
    const options = typeof prefixOrOptions === 'string'
        ? { prefix: prefixOrOptions }
        : { prefix: prefixOrOptions?.prefix };
    return (target) => {
        Reflect.defineMetadata(CONTROLLER_METADATA, options, target);
        // Auto-register in the DI container so the tool-bound instance is the same
        // singleton that receives lifecycle hooks (mirrors @Injectable behavior).
        const container = DIContainer.getInstance();
        if (!container.has(target)) {
            container.register(target);
        }
    };
}
/**
 * Get the tool-name prefix declared via @Controller, if any
 */
export function getControllerPrefix(target) {
    const metadata = Reflect.getMetadata(CONTROLLER_METADATA, target);
    return metadata?.prefix;
}
/**
 * Resource decorator - Marks a method as an MCP resource
 *
 * @example
 * ```typescript
 * @Resource({
 *   uri: 'db://users/schema',
 *   name: 'User Schema',
 *   description: 'Database schema for users',
 * })
 * async getUserSchema(context: ExecutionContext) {
 *   // Return schema
 * }
 * ```
 */
export function Resource(options) {
    return function (target, propertyKey, descriptor) {
        const existingResources = [...(Reflect.getMetadata(RESOURCE_METADATA, target.constructor) || [])];
        existingResources.push({
            methodName: String(propertyKey),
            options,
        });
        Reflect.defineMetadata(RESOURCE_METADATA, existingResources, target.constructor);
        return descriptor;
    };
}
/**
 * Prompt decorator - Marks a method as an MCP prompt
 *
 * @example
 * ```typescript
 * @Prompt({
 *   name: 'authentication-help',
 *   description: 'Help with authentication',
 * })
 * async authHelp(args: PromptArgs, context: ExecutionContext) {
 *   // Return prompt messages
 * }
 * ```
 */
export function Prompt(options) {
    return function (target, propertyKey, descriptor) {
        const existingPrompts = [...(Reflect.getMetadata(PROMPT_METADATA, target.constructor) || [])];
        existingPrompts.push({
            methodName: String(propertyKey),
            options,
        });
        Reflect.defineMetadata(PROMPT_METADATA, existingPrompts, target.constructor);
        return descriptor;
    };
}
/**
 * Extract tool definitions from a decorated class
 */
export function extractTools(target) {
    return Reflect.getMetadata(TOOL_METADATA, target) || [];
}
/**
 * Extract resource definitions from a decorated class
 */
export function extractResources(target) {
    return Reflect.getMetadata(RESOURCE_METADATA, target) || [];
}
/**
 * Extract prompt definitions from a decorated class
 */
export function extractPrompts(target) {
    return Reflect.getMetadata(PROMPT_METADATA, target) || [];
}
/**
 * Get widget metadata for a specific method
 */
export function getWidgetMetadata(target, methodName) {
    return Reflect.getMetadata(WIDGET_METADATA, target, methodName);
}
/**
 * Get guards metadata for a specific method
 */
export function getGuardsMetadata(target, methodName) {
    return Reflect.getMetadata(GUARDS_METADATA, target, methodName) || [];
}
/**
 * Initial tool decorator metadata key
 */
export const INITIAL_TOOL_METADATA = Symbol('initial_tool:metadata');
/**
 * InitialTool decorator - Marks a tool to be automatically called when the client starts
 *
 * @example
 * ```typescript
 * @Tool({ ... })
 * @InitialTool()
 * async init(input: InitInput, context: ExecutionContext) {
 *   // Implementation
 * }
 * ```
 */
export function InitialTool() {
    return function (target, propertyKey, descriptor) {
        Reflect.defineMetadata(INITIAL_TOOL_METADATA, true, target, String(propertyKey));
        return descriptor;
    };
}
/**
 * Get initial tool metadata for a specific method
 */
export function getInitialToolMetadata(target, methodName) {
    return Reflect.getMetadata(INITIAL_TOOL_METADATA, target, methodName) === true;
}
