// @nitrostack/core/dist/core/errors.js
/**
 * Base error class for MCP server errors
 */
export class McpError extends Error {
    code;
    statusCode;
    details;
    constructor(message, code, statusCode = 500, details) {
        super(message);
        this.code = code;
        this.statusCode = statusCode;
        this.details = details;
        this.name = 'McpError';
        Error.captureStackTrace(this, this.constructor);
    }
}
/**
 * Validation error
 */
export class ValidationError extends McpError {
    constructor(message, details) {
        super(message, 'VALIDATION_ERROR', 400, details);
        this.name = 'ValidationError';
    }
}
/**
 * Tool not found error
 */
export class ToolNotFoundError extends McpError {
    constructor(toolName) {
        super(`Tool '${toolName}' not found`, 'TOOL_NOT_FOUND', 404);
        this.name = 'ToolNotFoundError';
    }
}
/**
 * Resource not found error
 */
export class ResourceNotFoundError extends McpError {
    constructor(uri) {
        super(`Resource '${uri}' not found`, 'RESOURCE_NOT_FOUND', 404);
        this.name = 'ResourceNotFoundError';
    }
}
/**
 * Prompt not found error
 */
export class PromptNotFoundError extends McpError {
    constructor(promptName) {
        super(`Prompt '${promptName}' not found`, 'PROMPT_NOT_FOUND', 404);
        this.name = 'PromptNotFoundError';
    }
}
/**
 * Rate limit error
 */
export class RateLimitError extends McpError {
    constructor(message = 'Rate limit exceeded') {
        super(message, 'RATE_LIMIT_EXCEEDED', 429);
        this.name = 'RateLimitError';
    }
}
/**
 * Tool execution error
 */
export class ToolExecutionError extends McpError {
    constructor(toolName, originalError) {
        super(`Error executing tool '${toolName}': ${originalError.message}`, 'TOOL_EXECUTION_ERROR', 500, { originalError: originalError.message, stack: originalError.stack });
        this.name = 'ToolExecutionError';
    }
}
/**
 * Format error for response
 */
export function formatError(error) {
    if (error instanceof McpError) {
        return {
            code: error.code,
            message: error.message,
            details: error.details,
        };
    }
    return {
        code: 'INTERNAL_ERROR',
        message: error.message,
    };
}
