// @nitrostack/core/dist/core/jwt-module.js
import 'reflect-metadata';
/**
 * JWT Module - Enable JWT authentication in your MCP server
 *
 * Import this module to indicate your server uses JWT authentication.
 * Then use @UseGuards(JWTGuard) on your tools to protect them.
 *
 * @example
 * ```typescript
 * import { createServer, JWTModule } from 'nitrostack';
 *
 * const server = createServer({ ... });
 *
 * // Enable JWT authentication
 * server.use(JWTModule.forRoot({
 *   secretEnvVar: 'JWT_SECRET',
 *   expiresIn: '24h',
 * }));
 * ```
 */
export class JWTModule {
    static config = {
        secretEnvVar: 'JWT_SECRET',
        expiresIn: '24h',
    };
    /**
     * Configure JWT module for the application
     */
    static forRoot(config) {
        this.config = { ...this.config, ...config };
        return this.config;
    }
    /**
     * Get current JWT configuration
     */
    static getConfig() {
        return this.config;
    }
    /**
     * Get JWT secret from config or environment
     */
    static getSecret() {
        if (this.config.secret) {
            return this.config.secret;
        }
        if (this.config.secretEnvVar) {
            return process.env[this.config.secretEnvVar] || null;
        }
        return null;
    }
}
