// @nitrostack/core/dist/core/module.js
import 'reflect-metadata';
/**
 * Module metadata key
 */
export const MODULE_METADATA = Symbol('module:metadata');
/**
 * Module class (internal)
 */
class ModuleClass {
    metadata;
    constructor(metadata) {
        this.metadata = metadata;
    }
    /**
     * Get all controllers from this module
     */
    getControllers() {
        return this.metadata.controllers || [];
    }
    /**
     * Get module name
     */
    getName() {
        return this.metadata.name;
    }
    /**
     * Get module description
     */
    getDescription() {
        return this.metadata.description;
    }
}
/**
 * Module decorator - Defines a module with controllers, providers, imports
 *
 * @example
 * ```typescript
 * @Module({
 *   name: 'auth',
 *   description: 'Authentication module',
 *   controllers: [AuthController],
 * })
 * export class AuthModule {}
 * ```
 */
export function ModuleDecorator(metadata) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function (constructor) {
        // Store metadata on the class
        Reflect.defineMetadata(MODULE_METADATA, metadata, constructor);
        // Return the constructor with static method
        const enhanced = constructor;
        // Add static method
        Object.defineProperty(enhanced, 'getMetadata', {
            value: () => Reflect.getMetadata(MODULE_METADATA, constructor) || metadata,
            writable: false,
            enumerable: false,
            configurable: false,
        });
        return enhanced;
    };
}
// Export as "Module" for decorator usage
export { ModuleDecorator as Module };
/**
 * Create a module instance from a class
 */
export function createModule(moduleClass) {
    const metadata = Reflect.getMetadata(MODULE_METADATA, moduleClass);
    if (!metadata) {
        throw new Error(`Class ${moduleClass.name} is not decorated with @Module`);
    }
    return new ModuleClass(metadata);
}
/**
 * Check if a class is a module
 */
export function isModule(target) {
    return Reflect.hasMetadata(MODULE_METADATA, target);
}
/**
 * Get module metadata from a class
 */
export function getModuleMetadata(target) {
    return Reflect.getMetadata(MODULE_METADATA, target);
}
