// @nitrostack/core/dist/core/oauth-module.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var OAuthModule_1;
import 'reflect-metadata';
/**
 * OAuth Module - Enable OAuth 2.1 authentication in your MCP server
 *
 * This module provides:
 * - Protected Resource Metadata (RFC 9728)
 * - Token validation with audience binding (RFC 8707)
 * - Token introspection (RFC 7662)
 * - PKCE support (RFC 7636)
 *
 * Compatible with OpenAI Apps SDK and MCP specification.
 *
 * @example
 * ```typescript
 * import { McpApplicationFactory, OAuthModule } from 'nitrostack';
 * import { AppModule } from './app.module.js';
 *
 * @McpApp({
 *   module: AppModule,
 *   server: {
 *     name: 'OAuth MCP Server',
 *     version: '1.0.0',
 *   },
 * })
 * @Module({
 *   name: 'app',
 *   imports: [
 *     // Enable OAuth 2.1 authentication
 *     OAuthModule.forRoot({
 *       resourceUri: process.env.RESOURCE_URI!,
 *       authorizationServers: [process.env.AUTH_SERVER_URL!],
 *       scopesSupported: ['mcp:read', 'mcp:write', 'tools:execute'],
 *       tokenIntrospectionEndpoint: process.env.INTROSPECTION_ENDPOINT,
 *       tokenIntrospectionClientId: process.env.INTROSPECTION_CLIENT_ID,
 *       tokenIntrospectionClientSecret: process.env.INTROSPECTION_CLIENT_SECRET,
 *     }),
 *   ],
 * })
 * export class AppModule {}
 * ```
 */
import { Injectable, Inject } from './di/injectable.decorator.js';
import { NitroStackServer } from './server.js';
import { DiscoveryHttpServer } from './transports/discovery-http-server.js';
import { createAuthMiddleware } from '../auth/middleware.js';
import { validateToken as authValidateToken } from '../auth/token-validation.js';
let OAuthModule = class OAuthModule {
    static { OAuthModule_1 = this; }
    config;
    server;
    logger;
    static config = null;
    static discoveryInfo = null;
    discoveryServer = null;
    // Cache of upstream authorization-server metadata to avoid re-fetching on every request.
    // `data` is null for a (short-lived) negative cache entry.
    static discoveryMetadataCache = new Map();
    // In-flight fetches, keyed by auth server, to de-duplicate concurrent cache misses.
    static discoveryMetadataInflight = new Map();
    /** Max time to wait for an upstream metadata document before giving up. */
    static UPSTREAM_FETCH_TIMEOUT_MS = 4000;
    /** How long to remember that an upstream lookup failed, to avoid refetch storms. */
    static NEGATIVE_CACHE_TTL_MS = 30_000;
    async fetchUpstreamMetadata(authServer) {
        const now = Date.now();
        const cached = OAuthModule_1.discoveryMetadataCache.get(authServer);
        if (cached && cached.expires > now) {
            return cached.data;
        }
        // De-duplicate concurrent cache-miss fetches for the same authorization server.
        const inflight = OAuthModule_1.discoveryMetadataInflight.get(authServer);
        if (inflight) {
            return inflight;
        }
        const fetchPromise = (async () => {
            let metadata = null;
            for (const suffix of ['/.well-known/openid-configuration', '/.well-known/oauth-authorization-server']) {
                const controller = new AbortController();
                const timeout = setTimeout(() => controller.abort(), OAuthModule_1.UPSTREAM_FETCH_TIMEOUT_MS);
                try {
                    const response = await fetch(`${authServer}${suffix}`, { signal: controller.signal });
                    if (response.ok) {
                        metadata = (await response.json());
                        break;
                    }
                }
                catch (e) {
                    this.logger.debug(`OAuthModule: upstream metadata fetch failed for ${authServer}${suffix}`, {
                        error: e instanceof Error ? e.message : String(e),
                    });
                }
                finally {
                    clearTimeout(timeout);
                }
            }
            const positiveTtlMs = (this.config.tokenCacheSeconds ?? 300) * 1000;
            OAuthModule_1.discoveryMetadataCache.set(authServer, {
                data: metadata,
                expires: Date.now() + (metadata ? positiveTtlMs : OAuthModule_1.NEGATIVE_CACHE_TTL_MS),
            });
            return metadata;
        })();
        OAuthModule_1.discoveryMetadataInflight.set(authServer, fetchPromise);
        try {
            return await fetchPromise;
        }
        finally {
            OAuthModule_1.discoveryMetadataInflight.delete(authServer);
        }
    }
    buildBaseUrl(req) {
        // Prefer the configured, trusted resourceUri origin so advertised URLs are
        // not derived from client-controlled Host / X-Forwarded-Proto headers
        // (host-header injection). Fall back to request headers only if resourceUri
        // is missing or unparseable.
        try {
            if (this.config.resourceUri) {
                return new URL(this.config.resourceUri).origin;
            }
        }
        catch {
            // fall through to header-derived base
        }
        const reqHeaders = req?.headers ?? {};
        const rawHost = reqHeaders.host;
        const host = (Array.isArray(rawHost) ? rawHost[0] : rawHost) || 'localhost:3000';
        const rawProto = reqHeaders['x-forwarded-proto'];
        let proto = Array.isArray(rawProto) ? rawProto[0] : rawProto;
        if (!proto) {
            if (host.includes('localhost') || host.includes('127.0.0.1')) {
                proto = 'http';
            }
            else {
                proto = process.env.NODE_ENV === 'production' ? 'https' : 'http';
            }
        }
        return `${proto}://${host}`;
    }
    wellKnownHandler = async (req, res) => {
        const headers = {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type, Accept, Authorization',
        };
        const registrationEndpoint = this.isClientRegistrationEnabled()
            ? `${this.buildBaseUrl(req)}/oauth/v2/register`
            : undefined;
        try {
            const authServer = this.config.authorizationServers[0];
            const upstream = await this.fetchUpstreamMetadata(authServer);
            if (upstream) {
                // Clone before mutating so the cached object stays pristine
                const metadata = { ...upstream };
                // Inject registration_endpoint to satisfy strict client schema validation (Cursor/OpenAI)
                if (registrationEndpoint && !metadata.registration_endpoint) {
                    metadata.registration_endpoint = registrationEndpoint;
                }
                res.writeHead(200, headers);
                res.end(JSON.stringify(metadata));
                return;
            }
        }
        catch (err) {
            this.logger.debug('OAuthModule: failed to serve upstream authorization-server metadata; using fallback', {
                error: err instanceof Error ? err.message : String(err),
            });
        }
        // Fallback compliant with RFC 8414 / OIDC metadata schema
        const fallbackMetadata = {
            issuer: this.config.issuer || this.config.authorizationServers[0],
            authorization_endpoint: `${this.config.authorizationServers[0]}/oauth/v2/authorize`,
            token_endpoint: `${this.config.authorizationServers[0]}/oauth/v2/token`,
            introspection_endpoint: this.config.tokenIntrospectionEndpoint || `${this.config.authorizationServers[0]}/oauth/v2/introspect`,
            jwks_uri: this.config.jwksUri || `${this.config.authorizationServers[0]}/oauth/v2/keys`,
            response_types_supported: ['code'],
            grant_types_supported: ['authorization_code', 'client_credentials', 'refresh_token'],
            subject_types_supported: ['public'],
            id_token_signing_alg_values_supported: ['RS256'],
            code_challenge_methods_supported: ['S256'],
        };
        if (registrationEndpoint) {
            fallbackMetadata.registration_endpoint = registrationEndpoint;
        }
        res.writeHead(200, headers);
        res.end(JSON.stringify(fallbackMetadata));
    };
    /**
     * Whether the static Dynamic Client Registration endpoint is enabled.
     * Requires explicit opt-in AND a configured client id (never a literal default).
     */
    isClientRegistrationEnabled() {
        const enabled = this.config.enableClientRegistration === true
            || process.env.OAUTH_ENABLE_CLIENT_REGISTRATION === 'true';
        const clientId = this.config.staticClientId || process.env.OAUTH_CLIENT_ID;
        return enabled && !!clientId;
    }
    registrationHandler = async (req, res) => {
        const headers = {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type, Accept, Authorization',
        };
        if (req.method === 'OPTIONS') {
            res.writeHead(200, headers);
            res.end('');
            return;
        }
        // Registration is opt-in. Without explicit config + a configured client id we
        // must not hand out any client credentials.
        if (!this.isClientRegistrationEnabled()) {
            res.writeHead(404, headers);
            res.end(JSON.stringify({ error: 'not_found', error_description: 'Client registration is not enabled' }));
            return;
        }
        // Read and parse request body if not already parsed (for raw http.IncomingMessage)
        let body = {};
        if (req.body && typeof req.body === 'object') {
            body = req.body;
        }
        else {
            try {
                const buffers = [];
                for await (const chunk of req) {
                    buffers.push(chunk);
                }
                const data = Buffer.concat(buffers).toString();
                if (data) {
                    body = JSON.parse(data);
                }
            }
            catch (e) {
                this.logger.debug('OAuthModule: failed to parse client registration request body', {
                    error: e instanceof Error ? e.message : String(e),
                });
            }
        }
        // Client credentials come strictly from configuration / environment. No literal defaults.
        const clientId = this.config.staticClientId || process.env.OAUTH_CLIENT_ID;
        const clientSecret = this.config.staticClientSecret || process.env.OAUTH_CLIENT_SECRET || '';
        const responsePayload = {
            client_id: clientId,
            client_secret: clientSecret,
            client_id_issued_at: Math.floor(Date.now() / 1000),
            client_secret_expires_at: 0,
            grant_types: body.grant_types || ['authorization_code', 'refresh_token'],
            response_types: body.response_types || ['code'],
            token_endpoint_auth_method: body.token_endpoint_auth_method || (clientSecret ? 'client_secret_post' : 'none'),
            redirect_uris: body.redirect_uris || [],
        };
        res.writeHead(200, headers);
        res.end(JSON.stringify(responsePayload));
    };
    resourceMetadataHandler = (req, res) => {
        const headers = {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type, Accept, Authorization',
        };
        if (req?.method === 'OPTIONS') {
            res.writeHead(200, headers);
            res.end('');
            return;
        }
        // RFC 9728 - Protected Resource Metadata format
        const metadata = {
            resource: this.config.resourceUri,
            authorization_servers: this.config.authorizationServers,
        };
        // Add optional fields
        if (this.config.scopesSupported && this.config.scopesSupported.length > 0) {
            metadata.scopes_supported = this.config.scopesSupported;
        }
        res.writeHead(200, headers);
        res.end(JSON.stringify(metadata));
    };
    constructor(config, server, logger) {
        this.config = config;
        this.server = server;
        this.logger = logger;
        OAuthModule_1.config = config;
    }
    onModuleInit() {
        // Register the discovery endpoints
        // Handlers are now arrow functions, so no need to bind
    }
    /**
     * Get the preferred port for the OAuth discovery server
     * Priority: OAUTH_DISCOVERY_PORT > MCP_SERVER_PORT > PORT > config > 3005
     */
    getPreferredPort() {
        const parsePort = (value) => {
            if (!value)
                return null;
            const parsed = parseInt(value, 10);
            return Number.isNaN(parsed) ? null : parsed;
        };
        return (parsePort(process.env.OAUTH_DISCOVERY_PORT) ??
            parsePort(process.env.MCP_SERVER_PORT) ??
            parsePort(process.env.PORT) ??
            this.config.http?.port ??
            3005);
    }
    /**
     * Emit a clear, one-time startup log describing the current enforcement mode,
     * so a misconfigured or intentionally-open server is obvious in the logs.
     */
    logEnforcementMode() {
        const hasVerifier = !!this.config.jwksUri || !!this.config.tokenIntrospectionEndpoint;
        if (!this.config.required) {
            this.logger.warn('OAuthModule: authentication is NOT enforced (OAUTH_REQUIRED is not "true"). ' +
                'Protected endpoints are reachable without a token. ' +
                'Set OAUTH_REQUIRED=true (and configure JWKS_URI or INTROSPECTION_ENDPOINT) before deploying.');
            return;
        }
        if (!hasVerifier) {
            this.logger.warn('OAuthModule: OAUTH_REQUIRED is enabled but no token verifier is configured ' +
                '(set JWKS_URI or INTROSPECTION_ENDPOINT). The server will start, but ALL ' +
                'authenticated requests will be rejected (fail-closed) until a verifier is configured.');
            return;
        }
        this.logger.info('OAuthModule: authentication is enforced (OAUTH_REQUIRED=true) with a configured token verifier.');
    }
    async start() {
        this.logger.info('OAuthModule: start method called');
        this.logEnforcementMode();
        const transportType = this.server._transportType;
        const preferredPort = this.getPreferredPort();
        // Enable auto-retry by default in dev mode (stdio) to avoid port conflicts
        const autoRetry = process.env.OAUTH_DISCOVERY_AUTO_RETRY !== 'false';
        const serverOptions = {
            port: preferredPort,
            autoRetry: autoRetry,
            maxRetries: 50, // Try up to 50 ports to find an available one
        };
        if (transportType === 'stdio') {
            this.logger.info(`OAuthModule: Running in STDIO mode, starting DiscoveryHttpServer (preferred port: ${preferredPort})`);
            // In stdio mode, start a separate discovery server for OAuth endpoints
            this.discoveryServer = new DiscoveryHttpServer(serverOptions, this.logger);
            this.registerDiscoveryHandlers(this.discoveryServer);
            await this.discoveryServer.start();
            // Store the actual port for client discovery
            const actualPort = this.discoveryServer.getPort();
            this.updateDiscoveryInfo(actualPort);
            // Send notification to client about OAuth discovery info
            this.notifyClientAboutOAuth(actualPort);
        }
        else {
            this.logger.info(`OAuthModule: Running in ${transportType} mode, registering handlers with main server`);
            // In http or dual mode, register the handlers with the main server
            const httpTransport = this.server.getHttpTransport();
            if (httpTransport) {
                this.registerDiscoveryHandlers(httpTransport);
                // Retrieve Express app and register auth middleware BEFORE routes are compiled.
                // Enforcement is gated on `required`: when auth is not enforced, we leave the
                // discovery endpoints in place but do not mount the token-validating middleware.
                const app = httpTransport.getApp?.();
                if (app && this.config.required) {
                    const authMiddleware = createAuthMiddleware({
                        resourceUri: this.config.resourceUri,
                        authorizationServers: this.config.authorizationServers,
                        scopesSupported: this.config.scopesSupported,
                        tokenIntrospectionEndpoint: this.config.tokenIntrospectionEndpoint,
                        tokenIntrospectionClientId: this.config.tokenIntrospectionClientId,
                        tokenIntrospectionClientSecret: this.config.tokenIntrospectionClientSecret,
                        audience: this.config.audience,
                        issuer: this.config.issuer,
                        jwksUri: this.config.jwksUri,
                        tokenCacheSeconds: this.config.tokenCacheSeconds,
                    });
                    const basePath = this.config.http?.basePath || '/mcp';
                    // Express 4 `app.use(basePath, ...)` already matches basePath and all
                    // of its subpaths, so a separate `${basePath}/*` mount is redundant.
                    app.use(basePath, authMiddleware);
                    app.use('/sse', authMiddleware);
                    app.use('/mcp/messages', authMiddleware);
                    this.logger.info(`OAuthModule: Registered auth middleware on ${basePath} (and subpaths), /sse, and /mcp/messages`);
                }
                else if (app && !this.config.required) {
                    this.logger.info('OAuthModule: auth enforcement disabled (OAUTH_REQUIRED not set); skipping auth middleware');
                }
                // In HTTP mode, use the configured port
                this.updateDiscoveryInfo(preferredPort);
            }
            else {
                // Fallback: if httpTransport is not available, start a discovery server anyway
                // This handles edge cases where the transport setup fails or is delayed
                this.logger.warn(`OAuthModule: httpTransport not found for ${transportType} mode. Starting fallback DiscoveryHttpServer`);
                this.discoveryServer = new DiscoveryHttpServer(serverOptions, this.logger);
                this.registerDiscoveryHandlers(this.discoveryServer);
                await this.discoveryServer.start();
                const actualPort = this.discoveryServer.getPort();
                this.updateDiscoveryInfo(actualPort);
                this.notifyClientAboutOAuth(actualPort);
            }
        }
    }
    /**
     * Update the static discovery info
     */
    updateDiscoveryInfo(port) {
        OAuthModule_1.discoveryInfo = {
            enabled: true,
            discoveryPort: port,
            resourceUri: this.config.resourceUri,
            authorizationServers: this.config.authorizationServers,
            scopesSupported: this.config.scopesSupported,
        };
    }
    /**
     * Notify the client about OAuth configuration via stderr
     * This uses a JSON format that clients can parse from stderr
     */
    notifyClientAboutOAuth(port) {
        // Write OAuth discovery info to stderr in a parseable format
        // Clients can look for this JSON pattern to discover OAuth configuration
        const oauthInfo = {
            type: 'oauth_discovery',
            port: port,
            resourceUri: this.config.resourceUri,
            authorizationServers: this.config.authorizationServers,
            scopesSupported: this.config.scopesSupported,
            wellKnownEndpoints: {
                authorizationServer: `http://localhost:${port}/.well-known/oauth-authorization-server`,
                protectedResource: `http://localhost:${port}/.well-known/oauth-protected-resource`,
            },
        };
        // Output in a format that's easy to parse from stderr
        console.error(`[NITROSTACK_OAUTH]${JSON.stringify(oauthInfo)}[/NITROSTACK_OAUTH]`);
        this.logger.info(`OAuthModule: OAuth discovery info sent to client (port: ${port})`);
    }
    async stop() {
        if (this.discoveryServer) {
            await this.discoveryServer.stop();
            this.discoveryServer = null;
        }
        OAuthModule_1.discoveryInfo = null;
    }
    registerDiscoveryHandlers(server) {
        server.on('/.well-known/oauth-authorization-server', this.wellKnownHandler);
        server.on('/.well-known/oauth-protected-resource', this.resourceMetadataHandler);
        if (this.isClientRegistrationEnabled()) {
            server.on('/oauth/v2/register', this.registrationHandler);
        }
    }
    /**
     * Get the current OAuth discovery info
     * Returns null if OAuth is not configured or not started
     */
    static getDiscoveryInfo() {
        return this.discoveryInfo;
    }
    /**
     * Configure OAuth module for the application
     */
    static forRoot(config) {
        // Validate required fields
        if (!config.resourceUri) {
            throw new Error('OAuthModule: resourceUri is required');
        }
        if (!config.authorizationServers || config.authorizationServers.length === 0) {
            throw new Error('OAuthModule: at least one authorizationServer is required');
        }
        // Clone so env-derived defaults never mutate the caller's object.
        const resolved = { ...config };
        // Resolve defaults from environment variables
        if (!resolved.jwksUri && process.env.JWKS_URI) {
            resolved.jwksUri = process.env.JWKS_URI;
        }
        if (!resolved.audience) {
            resolved.audience = process.env.TOKEN_AUDIENCE || resolved.resourceUri;
        }
        if (!resolved.issuer && process.env.TOKEN_ISSUER) {
            resolved.issuer = process.env.TOKEN_ISSUER;
        }
        // Enforcement gate: explicit config wins, otherwise OAUTH_REQUIRED env, default false.
        resolved.required = config.required ?? (process.env.OAUTH_REQUIRED === 'true');
        // Eager URL format checks
        if (resolved.jwksUri) {
            try {
                new URL(resolved.jwksUri);
            }
            catch (err) {
                throw new Error(`OAuthModule: Invalid configuration value for jwksUri: ${err.message}`);
            }
        }
        this.config = resolved;
        return {
            module: OAuthModule_1,
            providers: [
                { provide: 'OAUTH_CONFIG', useValue: resolved }
            ],
        };
    }
    /**
     * Get current OAuth configuration
     */
    static getConfig() {
        return this.config;
    }
    /**
     * Whether OAuth authentication is enforced.
     *
     * Returns `false` when the module is unconfigured or `required` is not `true`,
     * in which case protected endpoints are reachable without a token.
     */
    static isAuthRequired() {
        return this.config?.required === true;
    }
    /**
     * Validate an access token
     *
     * Performs:
     * 1. Token introspection (if endpoint configured)
     * 2. Audience validation (RFC 8807)
     * 3. Issuer validation (if configured)
     * 4. Custom validation (if configured)
     */
    static async validateToken(token) {
        if (!this.config) {
            return { valid: false, error: 'OAuth module not configured' };
        }
        try {
            // Decode the header to check token type and provide helpful error message
            try {
                const headerPart = token.split('.')[0];
                const decodedHeader = JSON.parse(Buffer.from(headerPart, 'base64').toString());
                // Check if we received a JWE (encrypted) token instead of JWT
                if (decodedHeader.alg === 'dir' || decodedHeader.enc) {
                    return {
                        valid: false,
                        error: 'Received encrypted JWE token. MCP servers require unencrypted JWT access tokens. Check your OAuth provider application settings to ensure ID Token Encryption is disabled and that the "audience" parameter is being sent in authorization requests.'
                    };
                }
            }
            catch (headerError) {
                // If header decode fails, continue with normal validation
            }
            // If introspection or JWKS is configured, delegate to token-validation.ts pipeline
            if (this.config.tokenIntrospectionEndpoint || this.config.jwksUri) {
                const result = await authValidateToken(token, this.config);
                if (!result.valid || !result.introspection) {
                    return { valid: false, error: result.error || 'Invalid token' };
                }
                // Custom validation
                if (this.config.customValidation) {
                    const customValid = await this.config.customValidation(result.introspection);
                    if (!customValid) {
                        return { valid: false, error: 'Custom validation failed' };
                    }
                }
                return { valid: true, payload: result.introspection };
            }
            // No cryptographic verifier configured. Fail closed unless the caller has
            // explicitly opted into the insecure, signature-less decode fallback.
            if (!this.config.allowInsecureTokenDecode) {
                return {
                    valid: false,
                    error: 'No token validation method configured. Set tokenIntrospectionEndpoint or jwksUri (or enable allowInsecureTokenDecode for local development).',
                };
            }
            // Insecure fallback: decode token directly WITHOUT signature verification.
            // Only reachable when allowInsecureTokenDecode is true.
            const payload = this.decodeToken(token);
            if (!payload) {
                return { valid: false, error: 'Invalid token format' };
            }
            // Validate audience (RFC 8707 - critical for security). Fail closed when
            // the token carries no audience claim.
            if (!payload.aud) {
                return { valid: false, error: 'Token missing required audience (aud) claim' };
            }
            const audiences = Array.isArray(payload.aud) ? payload.aud : [payload.aud];
            if (!audiences.includes(this.config.audience)) {
                return {
                    valid: false,
                    error: `Token audience mismatch. Expected: ${this.config.audience}, Got: ${audiences.join(', ')}`,
                };
            }
            // Validate issuer
            if (this.config.issuer && payload.iss !== this.config.issuer) {
                return {
                    valid: false,
                    error: `Token issuer mismatch. Expected: ${this.config.issuer}, Got: ${payload.iss}`,
                };
            }
            // Check expiration
            const expiration = payload.exp;
            if (expiration && expiration < Date.now() / 1000) {
                return { valid: false, error: 'Token expired' };
            }
            // Custom validation
            if (this.config.customValidation) {
                const customValid = await this.config.customValidation(payload);
                if (!customValid) {
                    return { valid: false, error: 'Custom validation failed' };
                }
            }
            return { valid: true, payload };
        }
        catch (error) {
            return { valid: false, error: error instanceof Error ? error.message : String(error) };
        }
    }
    /**
     * Decode JWT token (without validation)
     * @private
     */
    static decodeToken(token) {
        try {
            const parts = token.split('.');
            if (parts.length !== 3)
                return null;
            // Convert base64url to base64 by replacing URL-safe characters
            let base64 = parts[1].replace(/-/g, '+').replace(/_/g, '/');
            // Add padding if necessary
            const padding = base64.length % 4;
            if (padding === 2) {
                base64 += '==';
            }
            else if (padding === 3) {
                base64 += '=';
            }
            const payload = JSON.parse(Buffer.from(base64, 'base64').toString('utf8'));
            return payload;
        }
        catch {
            return null;
        }
    }
};
OAuthModule = OAuthModule_1 = __decorate([
    Injectable(),
    __param(0, Inject('OAUTH_CONFIG')),
    __param(2, Inject('Logger')),
    __metadata("design:paramtypes", [Object, NitroStackServer, Object])
], OAuthModule);
export { OAuthModule };
