// @nitrostack/core/dist/core/resource.js
import { buildResourceReadContentsMeta } from './widget-mcp-meta.js';
/**
 * Resource class provides a clean abstraction for defining and serving resources
 */
export class Resource {
    definition;
    caches = new Map();
    subscribers = new Set();
    /** Widget template metadata; merged into resources/read contents[]._meta (e.g. MCP Apps ui.csp). */
    widgetReadMeta;
    constructor(definition) {
        this.definition = definition;
    }
    /**
     * Attach metadata from a UI component (OpenAI / MCP Apps) for ReadResource responses.
     */
    attachWidgetReadMeta(meta) {
        this.widgetReadMeta = meta;
    }
    getWidgetReadMeta() {
        return this.widgetReadMeta;
    }
    /**
     * Get resource URI
     */
    get uri() {
        return this.definition.uri;
    }
    /**
     * Get resource name
     */
    get name() {
        return this.definition.name;
    }
    /**
     * Get resource title (display name)
     */
    get title() {
        return this.definition.title;
    }
    /**
     * Get resource description
     */
    get description() {
        return this.definition.description;
    }
    /**
     * Get resource MIME type
     */
    get mimeType() {
        return this.definition.mimeType;
    }
    /**
     * Get resource size in bytes
     */
    get size() {
        return this.definition.size;
    }
    /**
     * Get resource annotations
     */
    get annotations() {
        return this.definition.annotations;
    }
    /**
     * Add a subscriber to this resource
     */
    subscribe(subscriberId) {
        this.subscribers.add(subscriberId);
    }
    /**
     * Remove a subscriber from this resource
     */
    unsubscribe(subscriberId) {
        this.subscribers.delete(subscriberId);
    }
    /**
     * Get all subscriber IDs
     */
    getSubscribers() {
        return Array.from(this.subscribers);
    }
    /**
     * Check if there are any subscribers
     */
    hasSubscribers() {
        return this.subscribers.size > 0;
    }
    /**
     * Fetch the resource content
     * @param context - Execution context
     * @param uri - Optional specific URI to fetch (defaults to resource.uri)
     */
    async fetch(context, uri) {
        const targetUri = uri || this.uri;
        const cacheable = this.definition.metadata?.cacheable;
        const cacheMaxAge = this.definition.metadata?.cacheMaxAge || 60000; // 1 minute default
        // Check cache for this specific URI
        const existingCache = this.caches.get(targetUri);
        if (cacheable && existingCache) {
            const age = Date.now() - existingCache.timestamp;
            if (age < cacheMaxAge) {
                context.logger.debug(`Serving cached resource: ${targetUri}`);
                return existingCache.content;
            }
        }
        context.logger.info(`Fetching resource: ${targetUri}`);
        try {
            const content = await this.definition.handler(targetUri, context);
            // Update cache for this specific URI
            if (cacheable) {
                this.caches.set(targetUri, {
                    content,
                    timestamp: Date.now(),
                });
            }
            context.logger.info(`Resource fetched successfully: ${targetUri}`);
            return content;
        }
        catch (error) {
            context.logger.error(`Error fetching resource: ${targetUri}`, { error: error instanceof Error ? error.message : String(error) });
            throw error;
        }
    }
    /**
     * Clear cached content for all URIs or a specific one
     */
    clearCache(uri) {
        if (uri) {
            this.caches.delete(uri);
        }
        else {
            this.caches.clear();
        }
    }
    /**
     * MCP Resource metadata structure (full spec compliance)
     */
    toMcpResource() {
        const resource = {
            uri: this.uri,
            name: this.name,
            description: this.description,
        };
        if (this.title)
            resource.title = this.title;
        if (this.mimeType)
            resource.mimeType = this.mimeType;
        if (this.size !== undefined)
            resource.size = this.size;
        if (this.annotations)
            resource.annotations = this.annotations;
        // Add mode-specific metadata if attached
        if (this.widgetReadMeta) {
            const meta = buildResourceReadContentsMeta(this.widgetReadMeta);
            if (meta) {
                resource._meta = meta;
            }
        }
        return resource;
    }
}
/**
 * Resource Template class for parameterized resources (RFC 6570)
 */
export class ResourceTemplate {
    definition;
    constructor(definition) {
        this.definition = definition;
    }
    get uriTemplate() {
        return this.definition.uriTemplate;
    }
    get name() {
        return this.definition.name;
    }
    get title() {
        return this.definition.title;
    }
    get description() {
        return this.definition.description;
    }
    get mimeType() {
        return this.definition.mimeType;
    }
    get annotations() {
        return this.definition.annotations;
    }
    /**
     * Convert to MCP protocol format
     */
    toMcpResourceTemplate() {
        const template = {
            uriTemplate: this.uriTemplate,
            name: this.name,
        };
        if (this.title)
            template.title = this.title;
        if (this.description)
            template.description = this.description;
        if (this.mimeType)
            template.mimeType = this.mimeType;
        if (this.annotations)
            template.annotations = this.annotations;
        return template;
    }
}
/**
 * Helper function to create a resource
 */
export function createResource(definition) {
    return new Resource(definition);
}
/**
 * Helper function to create a resource template
 */
export function createResourceTemplate(definition) {
    return new ResourceTemplate(definition);
}
