// @nitrostack/core/dist/core/server.js
import { Server as McpServer } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { SSEServerTransport } from '@modelcontextprotocol/sdk/server/sse.js';
import { CallToolRequestSchema, ListToolsRequestSchema, ListResourcesRequestSchema, ReadResourceRequestSchema, ListPromptsRequestSchema, GetPromptRequestSchema, ListResourceTemplatesRequestSchema, SubscribeRequestSchema, UnsubscribeRequestSchema, } from '@modelcontextprotocol/sdk/types.js';
import { ResourceTemplate, createResource } from './resource.js';
import { buildResourceReadContentsMeta } from './widget-mcp-meta.js';
import { getWidgetMimeType, isMcpAppMode, isOpenAiMode, getAppMode } from './app-mode.js';
import { createLogger } from './logger.js';
import { ToolExecutionError, ValidationError, ResourceNotFoundError, PromptNotFoundError } from './errors.js';
import { v4 as uuidv4 } from 'uuid';
import { isModule, getModuleMetadata } from './module.js';
import { buildController } from './builders.js';
import { DIContainer } from './di/container.js';
import { TaskManager, TaskContext, TaskNotFoundError, TaskAlreadyTerminalError, TaskAugmentationRequiredError, } from './task.js';
import { triggerLifecycleHook } from './lifecycle.js';
/** Optional Streamable HTTP session limits from environment variables. */
function getStreamableHttpEnvOptions() {
    const maxSessions = process.env.MCP_MAX_SESSIONS
        ? parseInt(process.env.MCP_MAX_SESSIONS, 10)
        : undefined;
    const sessionTimeout = process.env.MCP_SESSION_TIMEOUT_MS
        ? parseInt(process.env.MCP_SESSION_TIMEOUT_MS, 10)
        : undefined;
    return {
        ...(maxSessions !== undefined && !Number.isNaN(maxSessions) ? { maxSessions } : {}),
        ...(sessionTimeout !== undefined && !Number.isNaN(sessionTimeout) ? { sessionTimeout } : {}),
    };
}
/**
 * NitroStackServer - Main server class
 */
export class NitroStackServer {
    mcpServer;
    tools = new Map();
    resources = new Map();
    resourceTemplates = new Map();
    templateResources = new Map();
    prompts = new Map();
    modules = [];
    config;
    logger;
    stats = {
        toolCalls: 0,
        resourceReads: 0,
        promptExecutions: 0,
        errors: 0,
    };
    pendingComponentRegistrations = [];
    /** Transport type used by the server */
    _transportType;
    /** HTTP transport instance (when using http or dual mode) */
    _httpTransport;
    /** SDK legacy HTTP+SSE routes (GET /sse + POST /mcp/messages) attached once per Express app */
    _legacySseRoutesAttached = false;
    /** Active sessions for @modelcontextprotocol/sdk SSEServerTransport clients */
    legacySdkSseSessions = new Map();
    /** Task manager for MCP Tasks support */
    taskManager;
    /** Registered OS signal handlers for graceful shutdown (see enableShutdownHooks) */
    _shutdownSignalHandlers = [];
    /** Guards stop() so double signals / repeated calls don't re-run teardown */
    _stopping;
    constructor(config) {
        // Default config if not provided (e.g., when instantiated by DI container)
        this.config = config || {
            name: 'nitrostack-server',
            version: '1.0.0',
        };
        // Register itself in DI container so modules can inject the server.
        // NOTE: DIContainer is a process-wide singleton, so this assumes a single
        // NitroStackServer instance per process. Multiple instances would overwrite
        // each other's registration here.
        DIContainer.getInstance().registerValue(NitroStackServer, this);
        DIContainer.getInstance().registerValue('NitroStackServer', this);
        this.logger = createLogger({
            level: this.config.logging?.level || 'info',
            file: this.config.logging?.file,
            serviceName: this.config.name,
            enableConsole: false, // CRITICAL: Console disabled for MCP compatibility
        });
        // Initialize task manager for MCP Tasks support
        this.taskManager = new TaskManager({
            logger: this.logger,
            onStatusChange: (taskData) => {
                // Send notifications/tasks/status when task status changes
                this.sendTaskStatusNotification(taskData);
            },
        });
        this.mcpServer = new McpServer({
            name: this.config.name,
            version: this.config.version,
        }, NitroStackServer.mcpServerOptions);
        this.setupHandlersOn(this.mcpServer);
    }
    /** Shared MCP server constructor options (main + per legacy SSE session) */
    static mcpServerOptions = {
        capabilities: {
            tools: {
                listChanged: true,
            },
            resources: {
                subscribe: true,
                listChanged: true,
            },
            prompts: {
                listChanged: true,
            },
            tasks: {
                list: {},
                cancel: {},
                requests: {
                    tools: { call: {} },
                },
            },
        },
    };
    /**
     * Build a fresh MCP server instance wired with all of this application's
     * tool/resource/prompt handlers. Used for transports that require their own
     * server instance (legacy SSE sessions, official Streamable HTTP sessions),
     * since a single SDK server can only be connected to one transport at a time.
     */
    createConfiguredMcpServer() {
        const mcp = new McpServer({
            name: this.config.name,
            version: this.config.version,
        }, NitroStackServer.mcpServerOptions);
        this.setupHandlersOn(mcp);
        return mcp;
    }
    /**
     * SDK-compatible legacy HTTP+SSE: GET /sse (SSEServerTransport) and POST /mcp/messages?sessionId=.
     * Streamable HTTP stays on GET/POST /mcp.
     */
    async startLegacySdkSseSession(res, messagesPath) {
        try {
            const sessionMcp = this.createConfiguredMcpServer();
            const transport = new SSEServerTransport(messagesPath, res);
            const sessionId = transport.sessionId;
            this.legacySdkSseSessions.set(sessionId, { server: sessionMcp, transport });
            let closing = false;
            transport.onclose = async () => {
                if (closing) {
                    return;
                }
                closing = true;
                this.legacySdkSseSessions.delete(sessionId);
                try {
                    await sessionMcp.close();
                }
                catch {
                    // ignore
                }
            };
            transport.onerror = (err) => {
                this.logger.error('Legacy SDK SSE transport error', {
                    error: err instanceof Error ? err.message : String(err),
                });
            };
            await sessionMcp.connect(transport);
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            this.logger.error('Failed to start legacy SDK SSE session', { error: message });
            if (!res.headersSent) {
                res.status(500).end('Failed to establish SSE connection');
            }
        }
    }
    attachLegacySdkSseRoutes(app) {
        if (this._legacySseRoutesAttached) {
            return;
        }
        this._legacySseRoutesAttached = true;
        const LEGACY_SSE_PATH = '/sse';
        const LEGACY_MESSAGES_PATH = '/mcp/messages';
        app.get(LEGACY_SSE_PATH, async (_req, res) => {
            await this.startLegacySdkSseSession(res, LEGACY_MESSAGES_PATH);
        });
        app.post(LEGACY_MESSAGES_PATH, async (req, res) => {
            const sessionId = typeof req.query.sessionId === 'string' ? req.query.sessionId : undefined;
            if (!sessionId) {
                res.status(400).send('Missing sessionId query parameter');
                return;
            }
            const session = this.legacySdkSseSessions.get(sessionId);
            if (!session) {
                res.status(404).send('Unknown session');
                return;
            }
            try {
                await session.transport.handlePostMessage(req, res, req.body);
            }
            catch (error) {
                this.logger.error('Legacy SDK SSE POST failed', {
                    error: error instanceof Error ? error.message : String(error),
                });
                if (!res.headersSent) {
                    res.status(500).end('Failed to process message');
                }
            }
        });
    }
    attachLegacySdkSseIfNeeded(transport) {
        if (typeof transport.getApp === 'function') {
            this.attachLegacySdkSseRoutes(transport.getApp());
        }
        // Cursor opens GET /mcp without mcp-session-id; fall back to legacy SSE on that path.
        if (typeof transport.setLegacySseHandler === 'function') {
            transport.setLegacySseHandler(async (_req, res) => {
                await this.startLegacySdkSseSession(res, '/mcp/messages');
            });
        }
    }
    /**
     * Add a tool to the server
     */
    tool(tool) {
        this.tools.set(tool.name, tool);
        this.logger.info(`Tool registered: ${tool.name}`);
        // Auto-register component if attached
        if (tool.hasComponent()) {
            const component = tool.getComponent();
            // Track async component registration
            const registration = this.registerComponentResource(component).catch(err => {
                this.logger.error(`Failed to register component for tool ${tool.name}: ${err.message}`);
            });
            this.pendingComponentRegistrations.push(registration);
            this.logger.info(`Component auto-registered for tool: ${tool.name} -> ${component.getResourceUri()}`);
        }
        return this;
    }
    /**
     * Register a component as an MCP resource
     */
    async registerComponentResource(component) {
        // Compile component
        await component.compile();
        // Create resource for component
        const resource = createResource({
            uri: component.getResourceUri(),
            name: component.name,
            description: component.description || `UI component for ${component.name}`,
            mimeType: getWidgetMimeType(),
            handler: async (uri, context) => {
                context.logger.info(`Serving component: ${uri}`);
                // In production, serve the bundled HTML file if available
                if (process.env.NODE_ENV === 'production' || process.env.NODE_ENV === 'prod') {
                    try {
                        // Check if we have a bundled file for this component
                        // The component ID usually matches the widget output name
                        const widgetId = component.id;
                        // We need to find where the widgets are located relative to the running server
                        // In production, we expect them in src/widgets/out or dist/widgets/out
                        // Try to find the bundled file
                        const fs = await import('fs');
                        const path = await import('path');
                        // Possible locations for bundled widgets
                        const possiblePaths = [
                            path.join(process.cwd(), 'src/widgets/out', `${widgetId}.html`),
                            path.join(process.cwd(), 'dist/widgets/out', `${widgetId}.html`),
                            path.join(process.cwd(), 'widgets/out', `${widgetId}.html`)
                        ];
                        for (const p of possiblePaths) {
                            if (fs.existsSync(p)) {
                                const html = fs.readFileSync(p, 'utf-8');
                                return {
                                    type: 'text',
                                    data: html
                                };
                            }
                        }
                        context.logger.warn(`Bundled widget not found for ${widgetId}, falling back to default bundle`);
                    }
                    catch (error) {
                        context.logger.error(`Error serving bundled widget: ${error}`);
                    }
                }
                return {
                    type: 'text',
                    data: component.getBundle(),
                };
            },
        });
        const metadata = component.getResourceMetadata();
        if (metadata && Object.keys(metadata).length > 0) {
            resource.attachWidgetReadMeta(metadata);
        }
        // Register resource
        this.resource(resource);
    }
    /**
     * Add a resource to the server
     */
    resource(resource) {
        // Check if URI is a template (contains {variable})
        if (resource.uri.includes('{') && resource.uri.includes('}')) {
            const template = new ResourceTemplate({
                uriTemplate: resource.uri,
                name: resource.name,
                description: resource.description,
                mimeType: resource.mimeType,
                annotations: resource.annotations,
            });
            this.resourceTemplates.set(template.uriTemplate, template);
            this.templateResources.set(template.uriTemplate, resource);
            this.logger.info(`Resource template registered: ${template.uriTemplate}`);
        }
        // Always keep in main resources map so it shows up in resources/list
        this.resources.set(resource.uri, resource);
        this.logger.info(`Resource registered: ${resource.uri}`);
        return this;
    }
    /**
     * Add a prompt to the server
     */
    prompt(prompt) {
        this.prompts.set(prompt.name, prompt);
        this.logger.info(`Prompt registered: ${prompt.name}`);
        return this;
    }
    /**
     * Add a resource template to the server
     */
    resourceTemplate(template) {
        this.resourceTemplates.set(template.uriTemplate, template);
        this.logger.info(`Resource template registered: ${template.uriTemplate}`);
        return this;
    }
    /**
     * Notify clients that the list of resources has changed
     */
    notifyResourcesListChanged() {
        try {
            // Send notification through the MCP server
            const mcpServerWithNotification = this.mcpServer;
            if (mcpServerWithNotification.notification) {
                mcpServerWithNotification.notification({ method: 'notifications/resources/list_changed' })
                    .catch(err => this.logger.error('Failed to send resources list changed notification', { error: err instanceof Error ? err.message : String(err) }));
            }
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            this.logger.error('Error sending resources list changed notification', { error: errorMessage });
        }
    }
    /**
     * Notify clients that the list of prompts has changed
     */
    notifyPromptsListChanged() {
        try {
            const mcpServerWithNotification = this.mcpServer;
            if (mcpServerWithNotification.notification) {
                mcpServerWithNotification.notification({ method: 'notifications/prompts/list_changed' })
                    .catch(err => this.logger.error('Failed to send prompts list changed notification', { error: err instanceof Error ? err.message : String(err) }));
            }
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            this.logger.error('Error sending prompts list changed notification', { error: errorMessage });
        }
    }
    /**
     * Notify clients that the list of tools has changed
     */
    notifyToolsListChanged() {
        try {
            const mcpServerWithNotification = this.mcpServer;
            if (mcpServerWithNotification.notification) {
                mcpServerWithNotification.notification({ method: 'notifications/tools/list_changed' })
                    .catch(err => this.logger.error('Failed to send tools list changed notification', { error: err instanceof Error ? err.message : String(err) }));
            }
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            this.logger.error('Error sending tools list changed notification', { error: errorMessage });
        }
    }
    /**
     * Notify subscribers that a resource has been updated
     */
    notifyResourceUpdated(uri) {
        try {
            const resource = this.resources.get(uri);
            if (!resource || !resource.hasSubscribers())
                return;
            const mcpServerWithNotification = this.mcpServer;
            if (mcpServerWithNotification.notification) {
                mcpServerWithNotification.notification({
                    method: 'notifications/resources/updated',
                    params: { uri }
                }).catch(err => this.logger.error('Failed to send resource updated notification', { error: err instanceof Error ? err.message : String(err) }));
            }
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            this.logger.error('Error sending resource updated notification', { error: errorMessage });
        }
    }
    /**
     * Register a module with all its controllers
     * Automatically extracts and registers all tools, resources, and prompts
     *
     * @example
     * ```typescript
     * server.module(AuthModule);
     * server.module(ProductsModule);
     * ```
     */
    module(moduleClass) {
        this.modules.push(moduleClass);
        // Check if it's a module
        if (!isModule(moduleClass)) {
            throw new Error(`Class ${moduleClass.name} is not decorated with @Module. Use @Module decorator.`);
        }
        // Get module metadata
        const metadata = getModuleMetadata(moduleClass);
        if (!metadata) {
            throw new Error(`Failed to get metadata for module ${moduleClass.name}`);
        }
        this.logger.info(`Registering module: ${metadata.name}`);
        // Register providers in DI so declared-but-uninjected providers are still
        // instantiated during start() (via instantiateAll) and run lifecycle hooks,
        // matching the McpApplicationFactory path.
        const providers = metadata.providers || [];
        for (const provider of providers) {
            if (typeof provider === 'function') {
                DIContainer.getInstance().register(provider);
            }
            else if (provider && typeof provider === 'object' && 'provide' in provider) {
                const token = provider.provide;
                if (provider.useValue !== undefined) {
                    DIContainer.getInstance().registerValue(token, provider.useValue);
                }
                else if (provider.useClass) {
                    DIContainer.getInstance().register(token, provider.useClass);
                }
            }
        }
        // Process all controllers in the module
        const controllers = metadata.controllers || [];
        for (const controller of controllers) {
            this.logger.info(`  Processing controller: ${controller.name}`);
            // Register the controller in DI so it is instantiated during start()
            // (via instantiateAll) and participates in lifecycle hooks, matching the
            // McpApplicationFactory path.
            DIContainer.getInstance().register(controller);
            // Build all tools, resources, and prompts from controller
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const { tools, resources, prompts } = buildController(controller);
            // Register tools
            tools.forEach(tool => {
                this.tool(tool);
                this.logger.info(`    ✓ Tool: ${tool.name}`);
            });
            // Register resources
            resources.forEach(resource => {
                this.resource(resource);
                this.logger.info(`    ✓ Resource: ${resource.uri}`);
            });
            // Register prompts
            prompts.forEach(prompt => {
                this.prompt(prompt);
                this.logger.info(`    ✓ Prompt: ${prompt.name}`);
            });
        }
        this.logger.info(`Module registered: ${metadata.name} (${controllers.length} controller(s))`);
        return this;
    }
    /**
     * Get server statistics
     */
    getStats() {
        return { ...this.stats };
    }
    /**
     * Create execution context
     */
    createContext(options) {
        return {
            logger: this.logger,
            requestId: uuidv4(),
            toolName: options?.toolName,
            metadata: options?.metadata || {},
        };
    }
    /**
     * Register MCP protocol handlers on the given server instance (main or per legacy SSE session).
     */
    setupHandlersOn(mcp) {
        // List tools
        mcp.setRequestHandler(ListToolsRequestSchema, async () => {
            this.logger.debug('Listing tools');
            const tools = await Promise.all(Array.from(this.tools.values()).map((tool) => tool.toMcpTool()));
            return {
                tools,
            };
        });
        // Call tool
        mcp.setRequestHandler(CallToolRequestSchema, async (request) => {
            const { name, arguments: args } = request.params;
            const tool = this.tools.get(name);
            if (!tool) {
                throw new ToolExecutionError(name, new Error('Tool not found'));
            }
            // ----------------------------------------------------------------
            // MCP Tasks: detect task-augmented requests
            // The client sends `task: { ttl?: number }` in params to request
            // async task execution.
            // ----------------------------------------------------------------
            const requestParams = request.params;
            const taskParam = requestParams['task'];
            const isTaskAugmented = taskParam !== undefined;
            // Enforce tool-level task support negotiation
            if (isTaskAugmented && tool.taskSupport === 'forbidden') {
                throw {
                    code: -32601,
                    message: `Tool '${name}' does not support task augmentation`,
                };
            }
            if (!isTaskAugmented && tool.taskSupport === 'required') {
                throw new TaskAugmentationRequiredError();
            }
            // Extract _meta from args if present and add to context metadata
            const argsRecord = (args || {});
            const { _meta, ...toolArgs } = argsRecord;
            const context = this.createContext({
                metadata: _meta,
                toolName: name
            });
            // ----------------------------------------------------------------
            // Task-augmented path: create task, run async, return immediately
            // ----------------------------------------------------------------
            if (isTaskAugmented) {
                const taskData = this.taskManager.createTask(taskParam, name);
                const taskId = taskData.taskId;
                // Attach a TaskContext to the execution context so handlers can
                // report progress and check for cancellation
                const taskContext = new TaskContext(this.taskManager, taskId);
                context.task = taskContext;
                // Execute the tool asynchronously (fire-and-forget)
                this.runTaskAsync(tool, args, context, taskId, name);
                // Return CreateTaskResult immediately
                return {
                    task: taskData,
                };
            }
            // ----------------------------------------------------------------
            // Normal synchronous path
            // ----------------------------------------------------------------
            try {
                // Pass original args (including _meta) to tool
                const result = await tool.execute(args, context);
                this.stats.toolCalls++;
                // Check if tool has a UI component
                const response = {
                    content: [
                        {
                            type: 'text',
                            text: typeof result === 'string' ? result : JSON.stringify(result, null, 2),
                        },
                    ],
                };
                // Add structuredContent and _meta if component is attached
                if (tool.hasComponent()) {
                    const component = tool.getComponent();
                    // 1. Get structuredContent (for model and widget)
                    const transformedData = await component.transformData(result, context);
                    response.structuredContent = transformedData;
                    // 2. Get _meta (for widget state, hidden from model)
                    const widgetMeta = (await component.getWidgetMeta(result, context)) || {};
                    response._meta = widgetMeta;
                    // 3. Add mode-specific metadata
                    const meta = response._meta;
                    if (isMcpAppMode()) {
                        // MCP Apps mode: Always ensure _meta.ui.resourceUri is present
                        if (!meta.ui) {
                            meta.ui = { resourceUri: component.getResourceUri() };
                        }
                        else if (typeof meta.ui === 'object' && meta.ui !== null && !('resourceUri' in meta.ui)) {
                            meta.ui.resourceUri = component.getResourceUri();
                        }
                    }
                    if (isOpenAiMode()) {
                        // OpenAI mode: Always ensure _meta['openai/outputTemplate'] is present
                        if (!meta['openai/outputTemplate']) {
                            meta['openai/outputTemplate'] = component.getResourceUri();
                        }
                    }
                    context.logger.info(`Tool response includes structured content for component: ${component.id}`);
                }
                return response;
            }
            catch (error) {
                this.stats.errors++;
                const errorMessage = error instanceof Error ? error.message : String(error);
                context.logger.error(`Tool execution failed: ${name}`, { error: errorMessage });
                const formattedError = error instanceof ValidationError || error instanceof ToolExecutionError
                    ? error
                    : new ToolExecutionError(name, error);
                return {
                    content: [
                        {
                            type: 'text',
                            text: `Error: ${formattedError.message}`,
                        },
                    ],
                    isError: true,
                };
            }
        });
        // ----------------------------------------------------------------
        // MCP Tasks: tasks/get
        // Returns current task status. Blocks until terminal if needed.
        // ----------------------------------------------------------------
        this.registerCustomHandler(mcp, 'tasks/get', async (params) => {
            const { taskId } = params;
            if (!taskId) {
                throw { code: -32602, message: 'Invalid params: taskId is required' };
            }
            try {
                return this.taskManager.getTask(taskId);
            }
            catch (err) {
                if (err instanceof TaskNotFoundError) {
                    throw { code: -32602, message: err.message };
                }
                throw { code: -32603, message: 'Internal error' };
            }
        });
        // ----------------------------------------------------------------
        // MCP Tasks: tasks/result
        // Blocks until the task is in a terminal state, then returns the
        // underlying tool call result or error.
        // ----------------------------------------------------------------
        this.registerCustomHandler(mcp, 'tasks/result', async (params) => {
            const { taskId } = params;
            if (!taskId) {
                throw { code: -32602, message: 'Invalid params: taskId is required' };
            }
            try {
                const { result, error } = await this.taskManager.getResult(taskId);
                if (error) {
                    // Re-throw the original error so the client gets the JSON-RPC error
                    throw error;
                }
                // Attach related-task metadata to the result
                const resultWithMeta = result;
                return {
                    ...resultWithMeta,
                    _meta: {
                        ...(resultWithMeta?._meta || {}),
                        'io.modelcontextprotocol/related-task': { taskId },
                    },
                };
            }
            catch (err) {
                if (err instanceof TaskNotFoundError) {
                    throw { code: -32602, message: err.message };
                }
                // If err already looks like a JSON-RPC error, re-throw as-is
                if (err && typeof err === 'object' && 'code' in err) {
                    throw err;
                }
                throw { code: -32603, message: 'Internal error' };
            }
        });
        // ----------------------------------------------------------------
        // MCP Tasks: tasks/list
        // Lists tasks with cursor-based pagination.
        // ----------------------------------------------------------------
        this.registerCustomHandler(mcp, 'tasks/list', async (params) => {
            const { cursor } = (params || {});
            try {
                return this.taskManager.listTasks(cursor);
            }
            catch (err) {
                if (err instanceof TaskNotFoundError) {
                    // Invalid cursor
                    throw { code: -32602, message: 'Invalid params: invalid cursor' };
                }
                throw { code: -32603, message: 'Internal error' };
            }
        });
        // ----------------------------------------------------------------
        // MCP Tasks: tasks/cancel
        // Cancels an active task.
        // ----------------------------------------------------------------
        this.registerCustomHandler(mcp, 'tasks/cancel', async (params) => {
            const { taskId } = params;
            if (!taskId) {
                throw { code: -32602, message: 'Invalid params: taskId is required' };
            }
            try {
                return this.taskManager.cancelTask(taskId);
            }
            catch (err) {
                if (err instanceof TaskNotFoundError) {
                    throw { code: -32602, message: err.message };
                }
                if (err instanceof TaskAlreadyTerminalError) {
                    throw { code: -32602, message: err.message };
                }
                throw { code: -32603, message: 'Internal error' };
            }
        });
        // List resources
        mcp.setRequestHandler(ListResourcesRequestSchema, async () => {
            this.logger.debug('Listing resources');
            return {
                resources: Array.from(this.resources.values()).map((resource) => resource.toMcpResource()),
            };
        });
        // Read resource
        mcp.setRequestHandler(ReadResourceRequestSchema, async (request) => {
            const { uri } = request.params;
            let resource = this.resources.get(uri);
            // If not found as a static resource, check templates
            if (!resource) {
                for (const [uriTemplate, templateResource] of this.templateResources.entries()) {
                    // Robust template matching: 
                    // 1. Escape regex special characters except for the template variables
                    // 2. Replace {var} placeholders with a capturing group ([^/]+)
                    const escapedTemplate = uriTemplate.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
                    const regexStr = escapedTemplate.replace(/\\\{(.+?)\\\}/g, '([^/]+)');
                    const regex = new RegExp(`^${regexStr}$`);
                    if (regex.test(uri)) {
                        resource = templateResource;
                        break;
                    }
                }
            }
            if (!resource) {
                throw new ResourceNotFoundError(uri);
            }
            const context = this.createContext();
            try {
                const content = await resource.fetch(context, uri);
                let responseContent;
                switch (content.type) {
                    case 'text':
                        responseContent = {
                            uri: uri,
                            mimeType: resource.mimeType || 'text/plain',
                            text: content.data,
                        };
                        break;
                    case 'binary':
                        responseContent = {
                            uri: uri,
                            mimeType: resource.mimeType || 'application/octet-stream',
                            blob: content.data.toString('base64'),
                        };
                        break;
                    case 'json':
                        responseContent = {
                            uri: uri,
                            mimeType: resource.mimeType || 'application/json',
                            text: JSON.stringify(content.data, null, 2),
                        };
                        break;
                    default:
                        // Fallback: if content doesn't match ResourceContent shape, treat as JSON
                        responseContent = {
                            uri: uri,
                            mimeType: resource.mimeType || 'application/json',
                            text: JSON.stringify(content, null, 2),
                        };
                }
                const contentsMeta = buildResourceReadContentsMeta(resource.getWidgetReadMeta?.());
                if (contentsMeta) {
                    responseContent._meta = contentsMeta;
                }
                this.stats.resourceReads++;
                return {
                    contents: [responseContent],
                };
            }
            catch (error) {
                const errorMessage = error instanceof Error ? error.message : String(error);
                context.logger.error(`Resource fetch failed: ${uri}`, { error: errorMessage });
                throw error;
            }
        });
        // List resource templates (RFC 6570 URI templates)
        mcp.setRequestHandler(ListResourceTemplatesRequestSchema, async () => {
            this.logger.debug('Listing resource templates');
            return {
                resourceTemplates: Array.from(this.resourceTemplates.values()).map((template) => template.toMcpResourceTemplate()),
            };
        });
        // Subscribe to resource updates
        mcp.setRequestHandler(SubscribeRequestSchema, async (request) => {
            const { uri } = request.params;
            const resource = this.resources.get(uri);
            if (!resource) {
                throw new ResourceNotFoundError(uri);
            }
            // Use request ID as subscriber ID (in real implementation, use session ID)
            const subscriberId = uuidv4();
            resource.subscribe(subscriberId);
            this.logger.info(`Subscribed to resource: ${uri}`, { subscriberId });
            return {};
        });
        // Unsubscribe from resource updates
        mcp.setRequestHandler(UnsubscribeRequestSchema, async (request) => {
            const { uri } = request.params;
            const resource = this.resources.get(uri);
            if (!resource) {
                throw new ResourceNotFoundError(uri);
            }
            // Note: In a full implementation, track subscriber IDs per session
            this.logger.info(`Unsubscribe request for resource: ${uri}`);
            return {};
        });
        // List prompts
        mcp.setRequestHandler(ListPromptsRequestSchema, async () => {
            this.logger.debug('Listing prompts');
            return {
                prompts: Array.from(this.prompts.values()).map((prompt) => prompt.toMcpPrompt()),
            };
        });
        // Get prompt
        mcp.setRequestHandler(GetPromptRequestSchema, async (request) => {
            const { name, arguments: args } = request.params;
            const prompt = this.prompts.get(name);
            if (!prompt) {
                throw new PromptNotFoundError(name);
            }
            const context = this.createContext();
            try {
                const result = await prompt.execute(args || {}, context);
                this.stats.promptExecutions++;
                // Transform messages to MCP protocol format
                // MCP expects content to be an object with type and text, not a plain string
                const messages = result.map((msg) => ({
                    role: msg.role,
                    content: {
                        type: 'text',
                        text: msg.content,
                    },
                }));
                return {
                    description: prompt.description,
                    messages,
                };
            }
            catch (error) {
                this.stats.errors++;
                const errorMessage = error instanceof Error ? error.message : String(error);
                context.logger.error(`Prompt execution failed: ${name}`, { error: errorMessage });
                throw error;
            }
        });
        // Error handler
        mcp.onerror = (error) => {
            const errorMessage = error instanceof Error ? error.message : String(error);
            this.logger.error('MCP Server error', { error: errorMessage });
            this.stats.errors++;
        };
    }
    /**
     * Start the server
     * Automatically determines the transport based on NODE_ENV
     *
     * Transport determination:
     * - MCP_TRANSPORT_TYPE env var takes precedence (for explicit control)
     * - NODE_ENV=development or unset → stdio mode
     * - NODE_ENV=production → dual mode (STDIO + HTTP)
     */
    async start() {
        // Check for explicit transport type override
        const explicitTransport = process.env.MCP_TRANSPORT_TYPE;
        // Determine if we're in development mode
        // On Windows, NODE_ENV might not be passed correctly, so we're more lenient
        const nodeEnv = process.env.NODE_ENV?.toLowerCase();
        const isDevelopment = nodeEnv === 'development' || nodeEnv === 'dev' || !nodeEnv;
        // Use explicit transport if set, otherwise infer from NODE_ENV
        const transportType = explicitTransport || (isDevelopment ? 'stdio' : 'dual');
        this._transportType = transportType;
        this.logger.debug(`NitroStackServer.start(): NODE_ENV=${process.env.NODE_ENV}, MCP_TRANSPORT_TYPE=${explicitTransport}, transportType=${transportType}`);
        // Resolve all modules so they (and their dependencies) are initialized in the DI container
        for (const moduleClass of this.modules) {
            DIContainer.getInstance().resolve(moduleClass);
        }
        // Eagerly instantiate all registered providers/controllers so their lifecycle
        // hooks fire even when they aren't injected anywhere (NestJS-style singletons).
        DIContainer.getInstance().instantiateAll(this.logger);
        // Call onModuleInit for currently resolved instances (NestJS: after DI resolution)
        const initializedInstances = new Set(DIContainer.getInstance().getInstances());
        await triggerLifecycleHook([...initializedInstances], 'onModuleInit');
        // If HTTP transport is needed (dual or http mode), set it up BEFORE calling module.start()
        // This allows modules like OAuthModule to register endpoints/middleware on the HTTP server
        if (transportType === 'dual' || transportType === 'http') {
            const port = parseInt(process.env.PORT || '3000');
            const host = process.env.HOST || 'localhost';
            // Create HTTP transport first (do not start listening yet)
            const { StreamableHttpTransport } = await import('./transports/streamable-http.js');
            const httpTransport = new StreamableHttpTransport({
                port: port,
                host: host,
                endpoint: '/mcp',
                enableSessions: transportType === 'http', // Sessions ONLY in pure http mode
                enableCors: process.env.ENABLE_CORS !== 'false',
                ...getStreamableHttpEnvOptions(),
            });
            // Delegate /mcp protocol handling to the official SDK transport: each
            // session gets its own configured MCP server built via this factory.
            httpTransport.setMcpServerFactory(() => this.createConfiguredMcpServer());
            // Set up tools callback and server config for documentation page
            httpTransport.setToolsCallback(async () => {
                const tools = await Promise.all(Array.from(this.tools.values()).map((tool) => tool.toMcpTool()));
                return tools;
            });
            httpTransport.setServerConfig({
                name: this.config.name,
                version: this.config.version,
                description: this.config.description,
            });
            // Store HTTP transport reference BEFORE modules start
            // This allows OAuthModule to register discovery endpoints and middleware
            this._httpTransport = httpTransport;
        }
        // Call start for all modules (e.g., OAuthModule to register discovery endpoints and middleware)
        // Now _httpTransport is available for OAuthModule to use
        for (const moduleClass of this.modules) {
            const moduleInstance = DIContainer.getInstance().resolve(moduleClass);
            if (moduleInstance.start) {
                await moduleInstance.start();
            }
        }
        // Init any instances registered during module.start() that missed the first pass
        const allInstances = DIContainer.getInstance().getInstances();
        const lateInstances = allInstances.filter((instance) => !initializedInstances.has(instance));
        if (lateInstances.length > 0) {
            await triggerLifecycleHook(lateInstances, 'onModuleInit');
        }
        // NestJS: onApplicationBootstrap runs after init, before the app accepts connections
        await triggerLifecycleHook(DIContainer.getInstance().getInstances(), 'onApplicationBootstrap');
        // Start HTTP listener and attach legacy SSE routes ONLY AFTER dynamic modules have finished starting
        if ((transportType === 'dual' || transportType === 'http') && this._httpTransport) {
            this.attachLegacySdkSseIfNeeded(this._httpTransport);
            await this._httpTransport.start();
        }
        // Now complete the transport setup (connect MCP server) using the determined transportType
        const port = parseInt(process.env.PORT || '3000');
        const host = process.env.HOST || 'localhost';
        await this.startWithTransport(transportType, {
            port,
            host,
            endpoint: '/mcp',
            enableCors: process.env.ENABLE_CORS !== 'false',
        });
    }
    /**
     * Start the server with specified transport
     * @param transportType - 'stdio', 'http', or 'dual' (default: 'stdio')
     * @param transportOptions - Transport-specific options
     */
    async startWithTransport(transportType = 'stdio', transportOptions) {
        this._transportType = transportType;
        try {
            // Wait for all component registrations to complete
            if (this.pendingComponentRegistrations.length > 0) {
                this.logger.info(`Waiting for ${this.pendingComponentRegistrations.length} component(s) to compile...`);
                await Promise.all(this.pendingComponentRegistrations);
                this.pendingComponentRegistrations = []; // Clear after completion
                this.logger.info('All components compiled and registered');
            }
            if (transportType === 'dual') {
                // DUAL transport: STDIO + HTTP SSE
                // STDIO: For direct MCP connections (dev tools, Claude Desktop)
                // HTTP SSE: For web-based clients and multiple concurrent connections
                // 1. Ensure the HTTP host is running (normally created by start()).
                // The host owns /mcp and delegates protocol handling to the official SDK
                // Streamable HTTP transport via the MCP server factory (one server per
                // session), so no manual message forwarding is required here.
                let httpTransport = this._httpTransport;
                if (!httpTransport) {
                    const { StreamableHttpTransport } = await import('./transports/streamable-http.js');
                    const transport = new StreamableHttpTransport({
                        port: transportOptions?.port || 3000,
                        host: transportOptions?.host || 'localhost',
                        endpoint: transportOptions?.endpoint || '/mcp',
                        enableCors: transportOptions?.enableCors !== false, // Enable CORS by default for web clients
                        ...getStreamableHttpEnvOptions(),
                    });
                    transport.setMcpServerFactory(() => this.createConfiguredMcpServer());
                    this.attachLegacySdkSseIfNeeded(transport);
                    await transport.start();
                    httpTransport = transport;
                    this._httpTransport = httpTransport;
                }
                // 2. Connect the primary MCP server via STDIO for direct connections.
                const stdioTransport = new StdioServerTransport();
                await this.mcpServer.connect(stdioTransport);
                this.logger.info(`${this.config.name} started successfully (DUAL MODE)`);
                this.logger.info(`✨ Mode: ${getAppMode().toUpperCase()} (via NITROSTACK_APP_MODE)`);
                this.logger.info(`📡 STDIO: Ready for direct MCP connections`);
                this.logger.info(`🌐 Streamable HTTP: http://${transportOptions?.host || 'localhost'}:${transportOptions?.port || 3000}${transportOptions?.endpoint || '/mcp'}`);
                this.logger.info(`🌐 Legacy SDK SSE: http://${transportOptions?.host || 'localhost'}:${transportOptions?.port || 3000}/sse`);
            }
            else if (transportType === 'http') {
                // HTTP-only transport (Streamable HTTP with SSE)
                // Reuse if already created by start()
                let httpTransport = this._httpTransport;
                if (!httpTransport) {
                    const { StreamableHttpTransport } = await import('./transports/streamable-http.js');
                    const transport = new StreamableHttpTransport({
                        port: transportOptions?.port || 3000,
                        host: transportOptions?.host || 'localhost',
                        endpoint: transportOptions?.endpoint || '/mcp',
                        enableCors: transportOptions?.enableCors || false,
                        ...getStreamableHttpEnvOptions(),
                    });
                    // Delegate /mcp protocol handling to the official SDK transport.
                    transport.setMcpServerFactory(() => this.createConfiguredMcpServer());
                    // Set up tools callback and server config for documentation page
                    transport.setToolsCallback(async () => {
                        const tools = await Promise.all(Array.from(this.tools.values()).map((tool) => tool.toMcpTool()));
                        return tools;
                    });
                    transport.setServerConfig({
                        name: this.config.name,
                        version: this.config.version,
                        description: this.config.description,
                    });
                    this.attachLegacySdkSseIfNeeded(transport);
                    // Start HTTP server
                    await transport.start();
                    httpTransport = transport;
                    this._httpTransport = httpTransport;
                }
                // The HTTP host owns /mcp and manages its own per-session MCP servers via
                // the factory; no direct mcpServer.connect() is needed here.
                this.logger.info(`${this.config.name} started successfully (HTTP SSE transport)`);
                this.logger.info(`✨ Mode: ${getAppMode().toUpperCase()} (via NITROSTACK_APP_MODE)`);
                this.logger.info(`🌐 Streamable HTTP: http://${transportOptions?.host || 'localhost'}:${transportOptions?.port || 3000}${transportOptions?.endpoint || '/mcp'}`);
                this.logger.info(`🌐 Legacy SDK SSE: http://${transportOptions?.host || 'localhost'}:${transportOptions?.port || 3000}/sse`);
            }
            else {
                // STDIO-only transport (default)
                const transport = new StdioServerTransport();
                await this.mcpServer.connect(transport);
                this.logger.info(`${this.config.name} started successfully (STDIO transport)`);
                this.logger.info(`✨ Mode: ${getAppMode().toUpperCase()} (via NITROSTACK_APP_MODE)`);
            }
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            this.logger.error('Failed to start server', { error: errorMessage });
            throw error;
        }
    }
    /**
     * Execute a tool asynchronously as a task.
     * Called from the tools/call handler when task-augmented.
     * Updates task status and stores the result for tasks/result.
     */
    async runTaskAsync(tool, args, context, taskId, toolName) {
        try {
            const result = await tool.execute(args, context);
            this.stats.toolCalls++;
            const response = {
                content: [
                    {
                        type: 'text',
                        text: typeof result === 'string' ? result : JSON.stringify(result, null, 2),
                    },
                ],
            };
            if (tool.hasComponent()) {
                const component = tool.getComponent();
                const transformedData = await component.transformData(result, context);
                response.structuredContent = transformedData;
                response._meta = (await component.getWidgetMeta(result, context)) || {};
                const meta = response._meta;
                if (isMcpAppMode()) {
                    // MCP Apps mode: Always ensure _meta.ui.resourceUri is present
                    if (!meta.ui) {
                        meta.ui = { resourceUri: component.getResourceUri() };
                    }
                    else if (typeof meta.ui === 'object' && meta.ui !== null && !('resourceUri' in meta.ui)) {
                        meta.ui.resourceUri = component.getResourceUri();
                    }
                }
                if (isOpenAiMode()) {
                    // OpenAI mode: Always ensure _meta['openai/outputTemplate'] is present
                    if (!meta['openai/outputTemplate']) {
                        meta['openai/outputTemplate'] = component.getResourceUri();
                    }
                }
            }
            this.taskManager.completeTask(taskId, response);
        }
        catch (error) {
            this.stats.errors++;
            const errorMessage = error instanceof Error ? error.message : String(error);
            context.logger.error(`Task tool execution failed: ${toolName}`, { error: errorMessage, taskId });
            // Check if due to cancellation
            const taskData = this.taskManager.hasTask(taskId)
                ? this.taskManager.getTask(taskId)
                : null;
            if (taskData?.status === 'cancelled') {
                return; // Already cancelled — don't overwrite status
            }
            const formattedError = error instanceof ValidationError || error instanceof ToolExecutionError
                ? error
                : new ToolExecutionError(toolName, error);
            this.taskManager.failTask(taskId, { code: -32603, message: formattedError.message }, formattedError.message);
        }
    }
    /**
     * Register a custom JSON-RPC handler for task-related methods.
     * The MCP SDK's setRequestHandler only works with pre-defined schemas,
     * so we hook into the transport-level message handling.
     */
    registerCustomHandler(mcp, method, handler) {
        // Store handler for use in message routing (dual/HTTP mode)
        this._customHandlers.set(method, handler);
        const mcpServerInternal = mcp;
        if (mcpServerInternal._requestHandlers) {
            mcpServerInternal._requestHandlers.set(method, async (req) => {
                return handler(req.params || {});
            });
        }
    }
    /** Custom JSON-RPC handlers (for task methods not covered by MCP SDK schemas) */
    _customHandlers = new Map();
    /**
     * Send a notifications/tasks/status notification to the client.
     */
    sendTaskStatusNotification(taskData) {
        try {
            const mcpServerWithNotification = this.mcpServer;
            if (mcpServerWithNotification.notification) {
                mcpServerWithNotification.notification({
                    method: 'notifications/tasks/status',
                    params: taskData,
                }).catch(err => this.logger.error('Failed to send task status notification', {
                    error: err instanceof Error ? err.message : String(err),
                }));
            }
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            this.logger.error('Error sending task status notification', { error: errorMessage });
        }
    }
    /**
     * Get the TaskManager instance (useful for advanced use cases)
     */
    getTaskManager() {
        return this.taskManager;
    }
    /**
     * Register OS signal handlers so the server shuts down gracefully, running the
     * NestJS-style shutdown lifecycle hooks (beforeApplicationShutdown /
     * onApplicationShutdown) with the received signal.
     *
     * Opt-in (like NestJS `enableShutdownHooks`) to avoid attaching process
     * listeners for consumers that manage their own shutdown. Safe to call more
     * than once; handlers are only attached once per signal.
     *
     * @param signals - Signals to listen for (default: SIGTERM, SIGINT)
     */
    enableShutdownHooks(signals = ['SIGTERM', 'SIGINT']) {
        for (const signal of signals) {
            if (this._shutdownSignalHandlers.some((h) => h.signal === signal)) {
                continue;
            }
            const handler = () => {
                // stop() is idempotent, so repeated signals are safe.
                this.stop(signal).catch((error) => {
                    const errorMessage = error instanceof Error ? error.message : String(error);
                    this.logger.error(`Error during ${signal} shutdown`, { error: errorMessage });
                });
            };
            process.on(signal, handler);
            this._shutdownSignalHandlers.push({ signal, handler });
        }
        return this;
    }
    /**
     * Remove any process signal handlers registered via enableShutdownHooks().
     */
    removeShutdownHooks() {
        for (const { signal, handler } of this._shutdownSignalHandlers) {
            process.removeListener(signal, handler);
        }
        this._shutdownSignalHandlers = [];
    }
    /**
     * Stop the server
     *
     * Idempotent: concurrent or repeated calls (e.g. from multiple OS signals)
     * share the same in-flight teardown instead of re-running lifecycle hooks or
     * double-closing transports.
     */
    async stop(signal) {
        if (!this._stopping) {
            this._stopping = this.doStop(signal);
        }
        return this._stopping;
    }
    async doStop(signal) {
        // Detach signal handlers first so nothing re-enters teardown mid-shutdown.
        this.removeShutdownHooks();
        const instances = DIContainer.getInstance().getInstances();
        try {
            // Call onModuleDestroy for all resolved instances
            await triggerLifecycleHook(instances, 'onModuleDestroy', { safe: true, logger: this.logger });
            // Call beforeApplicationShutdown for all resolved instances
            await triggerLifecycleHook(instances, 'beforeApplicationShutdown', { safe: true, logger: this.logger }, signal);
            // Call stop for all modules. A single failing module must not prevent the
            // remaining modules (or transport teardown) from stopping.
            for (const moduleClass of this.modules) {
                const moduleInstance = DIContainer.getInstance().resolve(moduleClass);
                if (moduleInstance.stop) {
                    try {
                        await moduleInstance.stop();
                    }
                    catch (error) {
                        const errorMessage = error instanceof Error ? error.message : String(error);
                        this.logger.error(`Error stopping module ${moduleClass.name}`, { error: errorMessage });
                    }
                }
            }
            // Destroy task manager (stops cleanup interval)
            this.taskManager.destroy();
            for (const { server: sessionMcp, transport: legacyTransport } of this.legacySdkSseSessions.values()) {
                try {
                    await legacyTransport.close();
                }
                catch {
                    // ignore
                }
                try {
                    await sessionMcp.close();
                }
                catch {
                    // ignore
                }
            }
            this.legacySdkSseSessions.clear();
            this._legacySseRoutesAttached = false;
            // Close HTTP transport if running in dual mode
            if (this._httpTransport) {
                await this._httpTransport.close();
                this._httpTransport = undefined;
            }
            // Close MCP server
            await this.mcpServer.close();
            this.logger.info('Server stopped');
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            this.logger.error('Error stopping server', { error: errorMessage });
            throw error;
        }
        finally {
            // NestJS: always run after connection close attempts, even if close threw
            await triggerLifecycleHook(instances, 'onApplicationShutdown', { safe: true, logger: this.logger }, signal);
        }
    }
    /**
     * Get the HTTP transport (for modules that need to register endpoints)
     */
    getHttpTransport() {
        return this._httpTransport;
    }
}
/**
 * Helper function to create a server
 */
export function createServer(config) {
    return new NitroStackServer(config);
}
