// @nitrostack/core/dist/core/tool.js
import { DIContainer } from './di/container.js';
import { isMcpAppMode, isOpenAiMode } from './app-mode.js';
import { mergeToolUiMeta } from './widget-mcp-meta.js';
export class Tool {
    name;
    title;
    description;
    inputSchema;
    outputSchema;
    annotations;
    invocation;
    visibility;
    examples;
    widget;
    outputTemplate;
    isInitial;
    /** Task support level for this tool */
    taskSupport;
    handler;
    guards;
    middlewares;
    interceptors;
    pipes;
    filters;
    component;
    constructor(options) {
        this.name = options.name;
        this.title = options.title;
        this.description = options.description;
        this.inputSchema = options.inputSchema;
        this.outputSchema = options.outputSchema;
        this.annotations = options.annotations;
        this.invocation = options.invocation;
        this.handler = options.handler;
        this.guards = options.guards || [];
        this.middlewares = options.middlewares || [];
        this.interceptors = options.interceptors || [];
        this.pipes = options.pipes || [];
        this.filters = options.filters || [];
        this.examples = options.examples;
        this.widget = options.widget;
        this.outputTemplate = options.outputTemplate;
        this.isInitial = options.isInitial;
        this.taskSupport = options.taskSupport ?? 'forbidden';
        this.visibility = options.visibility || 'visible';
    }
    /**
     * Execute the tool with full pipeline:
     * Exception Filters -> Guards -> Middleware -> Interceptors -> Pipes -> Handler
     */
    async execute(input, context) {
        const container = DIContainer.getInstance();
        try {
            // 1. Execute Guards
            for (const GuardClass of this.guards) {
                const guard = container.has(GuardClass)
                    ? container.resolve(GuardClass)
                    : new GuardClass();
                const canActivate = await guard.canActivate(context);
                if (!canActivate) {
                    throw new Error('Access denied by guard');
                }
            }
            // 2. Build Middleware Chain
            const middlewareChain = async (chainInput) => {
                let index = 0;
                const middlewareInstances = this.middlewares.map(M => container.has(M) ? container.resolve(M) : new M());
                const next = async () => {
                    if (index >= middlewareInstances.length) {
                        // 3. Build Interceptor Chain
                        return await this.executeWithInterceptors(chainInput, context);
                    }
                    const middleware = middlewareInstances[index++];
                    return await middleware.use(context, next);
                };
                return await next();
            };
            return await middlewareChain(input);
        }
        catch (error) {
            // Execute Exception Filters
            if (this.filters.length > 0) {
                for (const FilterClass of this.filters) {
                    const filter = container.has(FilterClass)
                        ? container.resolve(FilterClass)
                        : new FilterClass();
                    return await filter.catch(error, context);
                }
            }
            // Re-throw if no filters handled it
            throw error;
        }
    }
    /**
     * Execute with interceptor chain
     */
    async executeWithInterceptors(input, context) {
        const container = DIContainer.getInstance();
        let index = 0;
        const interceptorInstances = this.interceptors.map(I => container.has(I) ? container.resolve(I) : new I());
        const next = async () => {
            if (index >= interceptorInstances.length) {
                // 4. Execute Pipes, then Handler
                return await this.executeWithPipes(input, context);
            }
            const interceptor = interceptorInstances[index++];
            return await interceptor.intercept(context, next);
        };
        return await next();
    }
    /**
     * Execute pipes and then the handler
     */
    async executeWithPipes(input, context) {
        const container = DIContainer.getInstance();
        let transformedInput = input;
        // Execute Pipes
        for (const PipeClass of this.pipes) {
            const pipe = container.has(PipeClass)
                ? container.resolve(PipeClass)
                : new PipeClass();
            transformedInput = await pipe.transform(transformedInput, {
                type: 'body',
                metatype: undefined,
                data: undefined,
            });
        }
        // Finally, execute the actual handler
        return await this.handler(transformedInput, context);
    }
    /**
     * Attach a UI component to this tool
     */
    setComponent(component) {
        this.component = component;
    }
    /**
     * Check if tool has a component
     */
    hasComponent() {
        return this.component !== undefined;
    }
    /**
     * Get the attached component
     */
    getComponent() {
        return this.component;
    }
    /**
     * Check if schema is a Zod schema
     */
    isZodSchema(schema) {
        return schema && typeof schema === 'object' && '_def' in schema;
    }
    /**
     * Convert schema to JSON Schema format
     */
    async convertToJsonSchema(schema) {
        if (this.isZodSchema(schema)) {
            try {
                const zodToJsonSchemaModule = await import('zod-to-json-schema');
                const zodToJsonSchema = zodToJsonSchemaModule.zodToJsonSchema || zodToJsonSchemaModule.default;
                return zodToJsonSchema(schema, {
                    $refStrategy: 'none',
                    target: 'jsonSchema7',
                });
            }
            catch (_error) {
                // Silently fall back to permissive schema — console output is
                // disabled in MCP servers as it breaks the JSON-RPC stdio protocol.
                return { type: 'object', properties: {}, additionalProperties: true };
            }
        }
        return schema;
    }
    /**
     * Convert to MCP tool format
     */
    async toMcpTool() {
        // Convert input schema
        let jsonSchema = await this.convertToJsonSchema(this.inputSchema);
        // Ensure type is set to "object" if not present
        if (!jsonSchema.type) {
            jsonSchema = { ...jsonSchema, type: 'object' };
        }
        // Convert output schema if present and ensure type: "object"
        let outputJsonSchema;
        if (this.outputSchema) {
            const converted = await this.convertToJsonSchema(this.outputSchema);
            const { type: outputType, ...outputRest } = converted;
            outputJsonSchema = { ...outputRest, type: 'object' };
        }
        // Construct the final schema, ensuring type is "object"
        const { type: schemaType, ...restSchema } = jsonSchema;
        const finalSchema = {
            ...restSchema,
            type: 'object',
        };
        const mcpTool = {
            name: this.name,
            description: this.description,
            inputSchema: finalSchema,
        };
        // Add optional fields per MCP spec
        if (this.title) {
            mcpTool.title = this.title;
        }
        if (outputJsonSchema) {
            mcpTool.outputSchema = outputJsonSchema;
        }
        if (this.annotations) {
            mcpTool.annotations = this.annotations;
        }
        // Add task support level (MCP Tasks spec)
        if (this.taskSupport && this.taskSupport !== 'forbidden') {
            mcpTool.execution = { taskSupport: this.taskSupport };
        }
        // Initialize _meta if needed
        mcpTool._meta = mcpTool._meta || {};
        // Initial tool flag
        if (this.isInitial) {
            mcpTool._meta['tool/initial'] = true;
        }
        // Include examples if available
        if (this.examples) {
            mcpTool.examples = this.examples;
            // CRITICAL: Also store examples in _meta to ensure they're preserved
            // through MCP serialization/deserialization. The studio checks _meta
            // fields, and custom top-level fields may be stripped by the MCP SDK.
            mcpTool._meta['tool/examples'] = this.examples;
        }
        // Include widget if available (for backward compatibility)
        if (this.widget) {
            mcpTool.widget = this.widget;
        }
        // Include outputTemplate if available (for backward compatibility)
        if (this.outputTemplate) {
            mcpTool.outputTemplate = this.outputTemplate;
        }
        // CRITICAL: Store widget info in _meta to ensure it's preserved
        // through MCP serialization/deserialization. The studio checks _meta
        // fields, and custom top-level fields may be stripped by the MCP SDK.
        //
        // We support multiple formats for compatibility:
        // - MCP Apps spec: _meta.ui = { resourceUri: '...' }
        // - OpenAI Apps SDK: _meta['openai/outputTemplate'] = '...'
        // - Generic: _meta['ui/template'] = '...'
        // 1. Check if we have an attached component (preferred)
        if (this.component) {
            const resourceUri = this.component.getResourceUri();
            const componentMeta = this.component.getResourceMetadata();
            // Always include generic format for internal NitroStack use
            mcpTool._meta['ui/template'] = resourceUri;
            if (isOpenAiMode()) {
                // OpenAI mode: openai/* keys
                mcpTool._meta['openai/outputTemplate'] = resourceUri;
                // Also set top-level outputTemplate for backward compatibility with some legacy hosts
                mcpTool.outputTemplate = resourceUri;
                if (componentMeta) {
                    const widgetCsp = componentMeta['openai/widgetCSP'];
                    if (widgetCsp !== undefined) {
                        mcpTool._meta['openai/widgetCSP'] = widgetCsp;
                    }
                    const widgetDesc = componentMeta['openai/widgetDescription'];
                    if (widgetDesc !== undefined) {
                        mcpTool._meta['openai/widgetDescription'] = widgetDesc;
                    }
                    const widgetBorder = componentMeta['openai/widgetPrefersBorder'];
                    if (widgetBorder !== undefined) {
                        mcpTool._meta['openai/widgetPrefersBorder'] = widgetBorder;
                    }
                    const widgetDomain = componentMeta['openai/widgetDomain'];
                    if (widgetDomain !== undefined) {
                        mcpTool._meta['openai/widgetDomain'] = widgetDomain;
                    }
                }
            }
            if (isMcpAppMode()) {
                // MCP Apps spec: _meta.ui
                const uiMeta = mergeToolUiMeta(resourceUri, componentMeta);
                // Add visibility (MCP Apps mode)
                if (this.visibility) {
                    uiMeta['visibility'] = this.visibility;
                }
                mcpTool._meta['ui'] = uiMeta;
            }
        }
        // 2. Fallback to legacy widget/outputTemplate options
        else if (this.widget || this.outputTemplate) {
            // Prioritize widget.route over outputTemplate
            const widgetRoute = this.widget?.route || this.outputTemplate;
            if (widgetRoute) {
                mcpTool._meta['ui/template'] = widgetRoute;
                if (isMcpAppMode()) {
                    const uiMeta = { resourceUri: widgetRoute };
                    if (this.visibility) {
                        uiMeta['visibility'] = this.visibility;
                    }
                    mcpTool._meta['ui'] = uiMeta;
                }
                if (isOpenAiMode()) {
                    mcpTool._meta['openai/outputTemplate'] = widgetRoute;
                }
            }
        }
        // Add invocation status messages (OpenAI Apps SDK — only if OpenAI mode is active)
        if (this.invocation && isOpenAiMode()) {
            if (this.invocation.invoking) {
                mcpTool._meta['openai/toolInvocation/invoking'] = this.invocation.invoking;
            }
            if (this.invocation.invoked) {
                mcpTool._meta['openai/toolInvocation/invoked'] = this.invocation.invoked;
            }
        }
        return mcpTool;
    }
}
