// @nitrostack/core/dist/core/widget-mime.js
/**
 * MIME type for interactive MCP / ChatGPT UI resources (SEP-1865).
 * Hosts that load "MCP App" iframes reject plain `text/html`.
 */
export const MCP_APP_WIDGET_MIME_TYPE = 'text/html;profile=mcp-app';
/**
 * Legacy OpenAI Apps SDK / Skybridge HTML resources (when not using MCP App profile).
 */
export const OPENAI_SKYBRIDGE_WIDGET_MIME_TYPE = 'text/html+skybridge';
/** Plain HTML — what many OpenAI / ChatGPT widget hosts expect in resources/list. */
export const OPENAI_PLAIN_HTML_WIDGET_MIME_TYPE = 'text/html';
/** @alias Default NitroStack widget resource MIME (MCP App profile) */
export const WIDGET_MIME_TYPE = MCP_APP_WIDGET_MIME_TYPE;
/** @alias Same as {@link MCP_APP_WIDGET_MIME_TYPE} */
export const MCP_APPS_MIME_TYPE = MCP_APP_WIDGET_MIME_TYPE;
/** @alias Skybridge-only */
export const OPENAI_SKYBRIDGE_MIME_TYPE = OPENAI_SKYBRIDGE_WIDGET_MIME_TYPE;
/**
 * Widget HTML MIME for `resources/read` + tool metadata.
 *
 * - Default / `mcp-app` / unset: `text/html;profile=mcp-app` (MCP Apps, MCPJam Inspector)
 * - `openai`: `text/html+skybridge` (ChatGPT / OpenAI Apps SDK pattern; see repo `todo-openai-apps-sdk-example`)
 * - `legacy` / `plain` / `text/html`: plain `text/html` (hosts that reject skybridge)
 * - `skybridge`: same as `openai`
 *
 * Override per component with {@link ComponentMetadata.widgetResourceMime}.
 */
export function resolveWidgetResourceMimeType(explicit) {
    if (explicit) {
        return explicit;
    }
    const v = process.env.NITRO_WIDGET_RESOURCE_MIME?.trim().toLowerCase();
    if (v === 'openai' || v === 'skybridge' || v === 'html+skybridge') {
        return OPENAI_SKYBRIDGE_WIDGET_MIME_TYPE;
    }
    if (v === 'legacy' || v === 'text/html' || v === 'plain') {
        return OPENAI_PLAIN_HTML_WIDGET_MIME_TYPE;
    }
    return MCP_APP_WIDGET_MIME_TYPE;
}
