// @nitrostack/cli/dist/commands/generate-types.js
import fs from 'fs';
import path from 'path';
import { createHeader, createSuccessBox, NitroSpinner, spacer } from '../ui/branding.js';
/**
 * Generates TypeScript types from tool definitions
 */
export async function generateTypes(options = {}) {
    const projectRoot = process.cwd();
    const outputPath = options.output || path.join(projectRoot, 'src', 'types', 'generated-tools.ts');
    console.log(createHeader('Types', 'Generate TypeScript definitions'));
    const spinner = new NitroSpinner('Scanning for tool definitions...').start();
    // Find all *.tools.ts files
    const toolFiles = await findToolFiles(path.join(projectRoot, 'src'));
    if (toolFiles.length === 0) {
        spinner.fail('No tool files found');
        return;
    }
    spinner.update(`Found ${toolFiles.length} tool file(s)`);
    // Parse tools and generate types
    const types = await generateTypesFromFiles(toolFiles);
    // Write output file
    const outputDir = path.dirname(outputPath);
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }
    fs.writeFileSync(outputPath, types);
    spinner.succeed('Type generation complete');
    spacer();
    console.log(createSuccessBox('Types Generated', [
        `Location: ${path.relative(projectRoot, outputPath)}`,
        `Count:    ${toolFiles.length} tools processed`
    ]));
}
async function findToolFiles(dir) {
    const files = [];
    if (!fs.existsSync(dir)) {
        return files;
    }
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);
        if (entry.isDirectory() && !entry.name.includes('node_modules') && !entry.name.includes('dist')) {
            files.push(...await findToolFiles(fullPath));
        }
        else if (entry.isFile() && entry.name.endsWith('.tools.ts')) {
            files.push(fullPath);
        }
    }
    return files;
}
async function generateTypesFromFiles(files) {
    let output = `/**
 * Auto-generated types from tool definitions
 * DO NOT EDIT THIS FILE MANUALLY
 * 
 * Generated by: nitrostack generate-types
 * Generated at: ${new Date().toISOString()}
 */

// ==========================================
// Common Types
// ==========================================

export interface ToolInput<T = unknown> {
  [key: string]: T;
}

export interface ToolOutput<T = unknown> {
  [key: string]: T;
}

// ==========================================
// Tool-specific Types
// ==========================================

`;
    const toolTypes = new Map();
    for (const file of files) {
        const content = fs.readFileSync(file, 'utf-8');
        // Extract @Tool decorators
        const toolRegex = /@Tool\(\{[\s\S]*?name:\s*['"]([^'"]+)['"]\s*,[\s\S]*?\}\)/g;
        let match;
        while ((match = toolRegex.exec(content)) !== null) {
            const toolName = match[1];
            const decoratorContent = match[0];
            // Extract inputSchema (Zod schema)
            const inputSchemaMatch = decoratorContent.match(/inputSchema:\s*z\.object\(\{([\s\S]*?)\}\)/);
            // Extract example response
            const examplesMatch = decoratorContent.match(/examples:\s*\{[\s\S]*?response:\s*(\{[\s\S]*?\})\s*\}/);
            if (inputSchemaMatch || examplesMatch) {
                const pascalCaseName = toPascalCase(toolName);
                // Generate input type from Zod schema
                let inputType = 'Record<string, unknown>';
                if (inputSchemaMatch) {
                    inputType = generateTypeFromZodSchema(inputSchemaMatch[1]);
                }
                // Generate output type from example
                let outputType = 'unknown';
                if (examplesMatch) {
                    outputType = generateTypeFromExample(examplesMatch[1]);
                }
                toolTypes.set(toolName, { input: inputType, output: outputType });
                output += `// ${toolName}\n`;
                output += `export type ${pascalCaseName}Input = ${inputType};\n`;
                output += `export type ${pascalCaseName}Output = ${outputType};\n\n`;
            }
        }
    }
    // Add a union type of all tool names
    output += `// ==========================================\n`;
    output += `// Tool Name Union\n`;
    output += `// ==========================================\n\n`;
    output += `export type ToolName = ${Array.from(toolTypes.keys()).map(name => `'${name}'`).join(' | ')};\n\n`;
    // Add a mapped type for inputs and outputs
    output += `// ==========================================\n`;
    output += `// Mapped Types\n`;
    output += `// ==========================================\n\n`;
    output += `export interface ToolInputs {\n`;
    for (const [toolName] of toolTypes) {
        const pascalCase = toPascalCase(toolName);
        output += `  '${toolName}': ${pascalCase}Input;\n`;
    }
    output += `}\n\n`;
    output += `export interface ToolOutputs {\n`;
    for (const [toolName] of toolTypes) {
        const pascalCase = toPascalCase(toolName);
        output += `  '${toolName}': ${pascalCase}Output;\n`;
    }
    output += `}\n`;
    return output;
}
function toPascalCase(str) {
    return str
        .split(/[_-]/)
        .map(word => word.charAt(0).toUpperCase() + word.slice(1))
        .join('');
}
function generateTypeFromZodSchema(zodContent) {
    // Simple parser for basic Zod schemas
    const fields = [];
    const fieldRegex = /(\w+):\s*z\.(\w+)\(\)(?:\.optional\(\))?(?:\.describe\([^)]*\))?/g;
    let match;
    while ((match = fieldRegex.exec(zodContent)) !== null) {
        const fieldName = match[1];
        const zodType = match[2];
        const isOptional = zodContent.includes(`${fieldName}:`) && zodContent.includes('.optional()');
        let tsType = 'unknown';
        switch (zodType) {
            case 'string':
                tsType = 'string';
                break;
            case 'number':
                tsType = 'number';
                break;
            case 'boolean':
                tsType = 'boolean';
                break;
            case 'enum':
                tsType = 'string';
                break;
            case 'array':
                tsType = 'unknown[]';
                break;
            case 'object':
                tsType = 'Record<string, unknown>';
                break;
        }
        fields.push(`${fieldName}${isOptional ? '?' : ''}: ${tsType}`);
    }
    return fields.length > 0 ? `{\n  ${fields.join(';\n  ')};\n}` : 'Record<string, unknown>';
}
function generateTypeFromExample(exampleJson) {
    try {
        const cleaned = exampleJson.replace(/\/\/.*/g, '').trim();
        const obj = eval('(' + cleaned + ')');
        return generateTypeFromObject(obj);
    }
    catch (e) {
        return 'unknown';
    }
}
function generateTypeFromObject(obj, indent = '  ') {
    if (obj === null || obj === undefined)
        return 'unknown';
    if (Array.isArray(obj)) {
        if (obj.length === 0)
            return 'unknown[]';
        return `Array<${generateTypeFromObject(obj[0], indent + '  ')}>`;
    }
    if (typeof obj === 'object') {
        const fields = [];
        for (const [key, value] of Object.entries(obj)) {
            fields.push(`${key}: ${generateTypeFromObject(value, indent + '  ')}`);
        }
        return `{\n${indent}${fields.join(`;\n${indent}`)};\n${indent.slice(2)}}`;
    }
    return typeof obj;
}
