// @nitrostack/cli/dist/commands/upgrade.js
import chalk from 'chalk';
import { execSync } from 'child_process';
import path from 'path';
import fs from 'fs-extra';
import https from 'https';
import { createHeader, createBox, createSuccessBox, createErrorBox, NitroSpinner, log, spacer, nextSteps, NITRO_BANNER_FULL, showFooter } from '../ui/branding.js';
import { trackEvent, shutdownAnalytics } from '../analytics/posthog.js';
import { cloneSkillsRepo } from '../skills/clone.js';
import { discoverSkills } from '../skills/discover.js';
import { installSkills } from '../skills/installer.js';
import { AGENTS } from '../skills/detect-agents.js';
/**
 * Fetch a package's latest published version from NPM using standard https module.
 */
export function fetchLatestNpmVersion(packageName) {
    return new Promise((resolve, reject) => {
        const url = `https://registry.npmjs.org/${packageName}/latest`;
        const req = https.get(url, {
            headers: {
                'Accept': 'application/json',
                'User-Agent': 'nitrostack-cli-upgrade'
            },
            timeout: 5000
        }, (res) => {
            if (res.statusCode !== 200) {
                reject(new Error(`Registry responded with HTTP ${res.statusCode}`));
                return;
            }
            let data = '';
            res.on('data', (chunk) => {
                data += chunk;
            });
            // Handle stream errors
            res.on('error', (err) => {
                reject(new Error(`Response stream error: ${err.message}`));
            });
            res.on('end', () => {
                try {
                    const parsed = JSON.parse(data);
                    if (parsed.version) {
                        resolve(parsed.version);
                    }
                    else {
                        reject(new Error('Invalid response structure from NPM registry'));
                    }
                }
                catch (e) {
                    reject(new Error(`Failed to parse response: ${e.message}`));
                }
            });
        });
        req.on('error', (err) => {
            reject(err);
        });
        req.on('timeout', () => {
            req.destroy();
            reject(new Error('Request timed out'));
        });
    });
}
/**
 * Get the current installed version of @nitrostack/core from package.json
 */
function getCoreVersion(packageJsonPath) {
    if (!fs.existsSync(packageJsonPath)) {
        return null;
    }
    const packageJson = fs.readJSONSync(packageJsonPath);
    return packageJson.dependencies?.['@nitrostack/core'] ||
        packageJson.devDependencies?.['@nitrostack/core'] ||
        null;
}
/**
 * Parse version string to extract the actual numeric version.
 *
 * Strips any leading range operator (^ ~ >= <= > <) and drops the pre-release
 * suffix (e.g. `-beta.1`) so the dot-separated segments always parse to numbers
 * instead of producing `NaN` in `compareVersions`.
 */
function parseVersion(versionString) {
    return versionString.replace(/^[\^~>=<]+/, '').split('-')[0];
}
/**
 * Compare two version strings
 */
export function compareVersions(v1, v2) {
    const parts1 = parseVersion(v1).split('.').map(Number);
    const parts2 = parseVersion(v2).split('.').map(Number);
    for (let i = 0; i < Math.max(parts1.length, parts2.length); i++) {
        const p1 = parts1[i] || 0;
        const p2 = parts2[i] || 0;
        if (p1 > p2)
            return 1;
        if (p1 < p2)
            return -1;
    }
    return 0;
}
/**
 * Determine if a dependency refers to a local file or workspace link.
 */
export function isLocalDependency(version) {
    return version.startsWith('file:') ||
        version.startsWith('link:') ||
        version.startsWith('workspace:') ||
        version.startsWith('.') ||
        version.startsWith('/');
}
/**
 * Update all @nitrostack/* versions in package.json using dynamic npm registry fetch
 */
async function updatePackageJson(packageJsonPath, dryRun) {
    if (!fs.existsSync(packageJsonPath)) {
        return [];
    }
    const packageJson = fs.readJSONSync(packageJsonPath);
    const results = [];
    let hasChanges = false;
    const updateDeps = async (deps) => {
        if (!deps)
            return;
        const promises = Object.keys(deps).map(async (pkg) => {
            if (pkg.startsWith('@nitrostack/')) {
                const currentVersion = deps[pkg];
                // Skip local dependencies entirely
                if (isLocalDependency(currentVersion)) {
                    return;
                }
                try {
                    const latestVersion = await fetchLatestNpmVersion(pkg);
                    if (compareVersions(currentVersion, latestVersion) < 0) {
                        results.push({
                            location: path.basename(path.dirname(packageJsonPath)),
                            packageName: pkg,
                            previousVersion: currentVersion,
                            newVersion: `^${latestVersion}`,
                            upgraded: true,
                        });
                        deps[pkg] = `^${latestVersion}`;
                        hasChanges = true;
                    }
                }
                catch (error) {
                    console.warn(`\n⚠️  Skipped upgrade check for ${pkg}: ${error.message}`);
                }
            }
        });
        await Promise.all(promises);
    };
    await updateDeps(packageJson.dependencies);
    await updateDeps(packageJson.devDependencies);
    if (hasChanges && !dryRun) {
        fs.writeJSONSync(packageJsonPath, packageJson, { spaces: 2 });
    }
    return results;
}
/**
 * Run npm install silently
 */
function runNpmInstall(directory) {
    execSync('npm install', {
        cwd: directory,
        stdio: 'pipe',
    });
}
/**
 * Main upgrade command handler
 */
export async function upgradeCommand(options) {
    console.log(NITRO_BANNER_FULL);
    console.log(createHeader('Upgrade', 'Update @nitrostack packages to latest'));
    trackEvent('cli_command_invoked', {
        command: 'upgrade',
        options: Object.keys(options).filter(k => options[k] !== undefined),
    });
    const projectRoot = process.cwd();
    const rootPackageJsonPath = path.join(projectRoot, 'package.json');
    const widgetsPath = path.join(projectRoot, 'src', 'widgets');
    const widgetsPackageJsonPath = path.join(widgetsPath, 'package.json');
    // Validate project
    if (!fs.existsSync(rootPackageJsonPath)) {
        console.log(createErrorBox('Not a NitroStack Project', 'package.json not found'));
        process.exit(1);
    }
    const coreVersion = getCoreVersion(rootPackageJsonPath);
    if (!coreVersion) {
        console.log(createErrorBox('Not a NitroStack Project', '@nitrostack/core is not a dependency'));
        process.exit(1);
    }
    const dryRun = options.dryRun ?? false;
    if (dryRun) {
        spacer();
        log('Dry run mode - no changes will be made', 'warning');
    }
    spacer();
    log('Checking for package updates...', 'info');
    spacer();
    const allResults = [];
    // Upgrade root
    const rootSpinner = new NitroSpinner('Checking root package.json...').start();
    try {
        const results = await updatePackageJson(rootPackageJsonPath, dryRun);
        if (results.length > 0) {
            allResults.push(...results);
            rootSpinner.succeed(`Root: Found ${results.length} package update(s)`);
            if (!dryRun) {
                const installSpinner = new NitroSpinner('Installing dependencies...').start();
                runNpmInstall(projectRoot);
                installSpinner.succeed('Root dependencies installed');
            }
        }
        else {
            rootSpinner.info('Root: All @nitrostack packages are up to date (or local references)');
        }
    }
    catch (error) {
        rootSpinner.fail('Failed to upgrade root');
        console.error(error);
    }
    // Upgrade widgets if they exist
    if (fs.existsSync(widgetsPackageJsonPath)) {
        const widgetsSpinner = new NitroSpinner('Checking widgets package.json...').start();
        try {
            const results = await updatePackageJson(widgetsPackageJsonPath, dryRun);
            if (results.length > 0) {
                allResults.push(...results);
                widgetsSpinner.succeed(`Widgets: Found ${results.length} package update(s)`);
                if (!dryRun) {
                    const installSpinner = new NitroSpinner('Installing widget dependencies...').start();
                    runNpmInstall(widgetsPath);
                    installSpinner.succeed('Widget dependencies installed');
                }
            }
            else {
                widgetsSpinner.info('Widgets: All @nitrostack packages are up to date (or local references)');
            }
        }
        catch (error) {
            widgetsSpinner.fail('Failed to upgrade widgets');
            console.error(error);
        }
    }
    // Upgrade skills
    const rootPackageJson = fs.readJSONSync(rootPackageJsonPath);
    const localSkillsVersion = rootPackageJson.nitrostack?.skillsVersion || '0.0.0';
    let tempSkillsDir = null;
    const skillsSpinner = new NitroSpinner('Checking agent skills...').start();
    try {
        tempSkillsDir = await cloneSkillsRepo();
        const skillsPkgJsonPath = path.join(tempSkillsDir, 'package.json');
        let remoteSkillsVersion = '1.0.0';
        if (fs.existsSync(skillsPkgJsonPath)) {
            const skillsPkgJson = fs.readJSONSync(skillsPkgJsonPath);
            remoteSkillsVersion = skillsPkgJson.version || '1.0.0';
        }
        if (compareVersions(localSkillsVersion, remoteSkillsVersion) < 0) {
            if (dryRun) {
                allResults.push({
                    location: 'skills',
                    packageName: '@nitrostack/skills',
                    previousVersion: localSkillsVersion,
                    newVersion: remoteSkillsVersion,
                    upgraded: false,
                });
                skillsSpinner.info(`Skills: Updates available (${localSkillsVersion} → ${remoteSkillsVersion})`);
            }
            else {
                const skills = await discoverSkills(tempSkillsDir);
                await installSkills(AGENTS, skills, true, 'project', projectRoot);
                // Write new version to package.json
                const updatedPackageJson = fs.readJSONSync(rootPackageJsonPath);
                if (!updatedPackageJson.nitrostack) {
                    updatedPackageJson.nitrostack = {};
                }
                updatedPackageJson.nitrostack.skillsVersion = remoteSkillsVersion;
                fs.writeJSONSync(rootPackageJsonPath, updatedPackageJson, { spaces: 2 });
                allResults.push({
                    location: 'skills',
                    packageName: '@nitrostack/skills',
                    previousVersion: localSkillsVersion,
                    newVersion: remoteSkillsVersion,
                    upgraded: true,
                });
                skillsSpinner.succeed(`Skills: Upgraded from ${localSkillsVersion} to ${remoteSkillsVersion}`);
            }
        }
        else {
            skillsSpinner.info('Skills: All agent skills are up to date');
        }
    }
    catch (error) {
        skillsSpinner.fail('Failed to upgrade agent skills');
        console.error(error);
    }
    finally {
        if (tempSkillsDir) {
            try {
                fs.removeSync(tempSkillsDir);
            }
            catch {
                // best effort
            }
        }
    }
    // Summary
    spacer();
    if (allResults.length === 0) {
        console.log(createSuccessBox('Already Up to Date', [
            'All @nitrostack packages are already running their latest versions.',
        ]));
        trackEvent('cli_upgrade_completed', {
            packages_upgraded: 0,
            dry_run: dryRun,
            already_current: true,
        });
        await shutdownAnalytics();
        return;
    }
    // Unique packages upgraded
    const uniquePackages = Array.from(new Set(allResults.map(r => r.packageName)));
    const summaryItems = uniquePackages.map(pkg => {
        const result = allResults.find(r => r.packageName === pkg);
        return `${pkg}: ${parseVersion(result.previousVersion)} → ${parseVersion(result.newVersion)}`;
    });
    if (dryRun) {
        spacer();
        console.log(createBox([
            chalk.yellow.bold('Dry Run - Proposed Upgrades:'),
            ...summaryItems.map(item => `  • ${item}`),
            '',
            chalk.dim('No changes were made to your project.'),
            chalk.dim('Run without --dry-run to apply the upgrade.'),
        ], 'warning'));
    }
    else {
        console.log(createSuccessBox('Upgrade Complete', [
            ...summaryItems,
            '',
            chalk.dim(`Total updates across all packages: ${allResults.length}`)
        ]));
        nextSteps([
            'Review the changes in package.json',
            'Restart your development server',
            'Check docs.nitrostack.ai for migration guides',
        ]);
    }
    showFooter();
    trackEvent('cli_upgrade_completed', {
        packages_upgraded: allResults.length,
        dry_run: dryRun,
        already_current: false,
    });
    await shutdownAnalytics();
}
