// @nitrostack/core/dist/core/di/container.js
import 'reflect-metadata';
/**
 * Dependency Injection Container
 *
 * Simple DI container for managing service instances and dependencies.
 * Supports constructor injection and singleton pattern.
 */
export class DIContainer {
    static instance;
    providers = new Map();
    instances = new Map();
    constructor() { }
    /**
     * Get the singleton container instance
     */
    static getInstance() {
        if (!DIContainer.instance) {
            DIContainer.instance = new DIContainer();
        }
        return DIContainer.instance;
    }
    /**
     * Register a provider (class)
     * @param token - The token to register (usually the class itself)
     * @param provider - Optional provider class (defaults to token if it's a class)
     */
    register(token, provider) {
        this.providers.set(token, (provider || token));
    }
    /**
     * Register a value directly
     * @param token - The token to register
     * @param value - The value to associate with the token
     */
    registerValue(token, value) {
        this.instances.set(token, value);
    }
    /**
     * Resolve a dependency (get or create instance)
     * @param token - The token to resolve
     * @returns The resolved instance
     */
    resolve(token) {
        // Check if already instantiated
        if (this.instances.has(token)) {
            return this.instances.get(token);
        }
        // Get the provider
        const provider = this.providers.get(token) || token;
        // If provider is not a constructor function, throw error
        if (typeof provider !== 'function') {
            throw new Error(`Cannot resolve token "${String(token)}". No value or provider registered.`);
        }
        // Get constructor dependencies
        const dependencies = this.getDependencies(provider);
        // Resolve dependencies recursively
        const resolvedDependencies = dependencies.map((dep) => this.resolve(dep));
        // Create instance
        const ProviderClass = provider;
        const instance = new ProviderClass(...resolvedDependencies);
        // Cache instance
        this.instances.set(token, instance);
        return instance;
    }
    /**
     * Get constructor dependencies from metadata
     *
     * Priority:
     * 1. Explicit deps in @Injectable({ deps: [...] }) - works in ESM
     * 2. @Inject() parameter decorators
     * 3. design:paramtypes from TypeScript (may not work in ESM)
     */
    getDependencies(target) {
        // FIRST: Check for explicit deps in @Injectable({ deps: [...] })
        // This is the most reliable method for ESM compatibility
        const explicitDeps = Reflect.getMetadata('nitrostack:deps', target);
        if (explicitDeps && explicitDeps.length > 0) {
            return explicitDeps;
        }
        // SECOND: Get design:paramtypes metadata set by TypeScript
        // This may be empty in ESM if reflect-metadata wasn't loaded early enough
        const params = Reflect.getMetadata('design:paramtypes', target) || [];
        // Check for @Inject tokens which override the types
        const injectTokens = Reflect.getMetadata('nitrostack:inject', target) || [];
        // Merge: use inject tokens where specified, otherwise use the param type
        return params.map((param, index) => {
            return injectTokens[index] || param;
        });
    }
    /**
     * Get all currently instantiated objects in the container
     */
    getInstances() {
        return Array.from(this.instances.values());
    }
    /**
     * Eagerly instantiate every registered provider.
     *
     * The container resolves lazily, so declared-but-uninjected providers (and
     * controllers) would otherwise never be constructed and would miss lifecycle
     * hooks. This mirrors NestJS's eager singleton instantiation. A provider that
     * fails to resolve is logged (if a logger is supplied) and skipped so one bad
     * token does not abort startup.
     *
     * @param logger - Optional logger for resolution failures
     */
    instantiateAll(logger) {
        for (const token of this.providers.keys()) {
            if (this.instances.has(token)) {
                continue;
            }
            try {
                this.resolve(token);
            }
            catch (error) {
                const errMsg = error instanceof Error ? error.message : String(error);
                logger?.error(`Failed to instantiate provider "${String(token)}": ${errMsg}`, { error });
            }
        }
    }
    /**
     * Clear all registrations (useful for testing)
     */
    clear() {
        this.providers.clear();
        this.instances.clear();
    }
    /**
     * Check if a token is registered
     * @param token - The token to check
     */
    has(token) {
        return this.providers.has(token) || this.instances.has(token);
    }
}
