// @nitrostack/core/dist/core/di/injectable.decorator.js
import 'reflect-metadata';
import { DIContainer } from './container.js';
const INJECTABLE_KEY = 'nitrostack:injectable';
const INJECT_KEY = 'nitrostack:inject';
const DEPS_KEY = 'nitrostack:deps';
/**
 * Marks a class as injectable (can be used with DI)
 *
 * @example Basic usage (relies on TypeScript's emitDecoratorMetadata):
 * ```typescript
 * @Injectable()
 * export class UserService {
 *   constructor(private db: DatabaseService) {}
 * }
 * ```
 *
 * @example With explicit deps (recommended for ESM compatibility):
 * ```typescript
 * @Injectable({ deps: [DatabaseService] })
 * export class UserService {
 *   constructor(private db: DatabaseService) {}
 * }
 * ```
 */
export function Injectable(options) {
    return (target) => {
        Reflect.defineMetadata(INJECTABLE_KEY, true, target);
        // Store explicit deps if provided (for ESM compatibility)
        if (options?.deps && options.deps.length > 0) {
            Reflect.defineMetadata(DEPS_KEY, options.deps, target);
        }
        // Auto-register in DI container
        const container = DIContainer.getInstance();
        if (!container.has(target)) {
            container.register(target);
        }
    };
}
/**
 * Inject a specific token (optional - for custom injection)
 *
 * @example
 * ```typescript
 * constructor(@Inject('DATABASE_CONFIG') private config: DatabaseConfig) {}
 * ```
 */
export function Inject(token) {
    return (target, propertyKey, parameterIndex) => {
        // Clone to avoid mutating a parent class's inject-token array when subclassing.
        const existingTokens = [...(Reflect.getMetadata(INJECT_KEY, target) || [])];
        existingTokens[parameterIndex] = token;
        Reflect.defineMetadata(INJECT_KEY, existingTokens, target);
    };
}
/**
 * Check if a class is injectable
 */
export function isInjectable(target) {
    return Reflect.getMetadata(INJECTABLE_KEY, target) === true;
}
/**
 * Get inject tokens for constructor parameters
 */
export function getInjectTokens(target) {
    return Reflect.getMetadata(INJECT_KEY, target) || [];
}
