// @nitrostack/core/dist/auth/simple-jwt.js
import jwt from 'jsonwebtoken';
import { unwrapSecret } from './secure-secret.js';
/**
 * Create Simple JWT authentication middleware
 *
 * @example
 * ```typescript
 * const server = createServer({...});
 *
 * // Simple JWT auth (no OAuth complexity!)
 * server.app.use('/mcp', createSimpleJWTAuth({
 *   secret: process.env.JWT_SECRET!,
 *   audience: 'my-mcp-server',
 *   issuer: 'my-app',
 * }));
 *
 * server.start();
 * ```
 */
export function createSimpleJWTAuth(config) {
    const algorithm = config.algorithm || 'HS256';
    return async (req, res, next) => {
        try {
            // 1. Extract token from Authorization header
            const authHeader = req.headers.authorization;
            if (!authHeader || !authHeader.startsWith('Bearer ')) {
                return res.status(401).json({
                    error: 'unauthorized',
                    message: 'Missing or invalid Authorization header. Use: Authorization: Bearer <token>',
                });
            }
            const token = authHeader.substring(7); // Remove 'Bearer '
            // 2. Verify JWT
            const verifyOptions = {
                algorithms: [algorithm],
            };
            if (config.audience) {
                verifyOptions.audience = config.audience;
            }
            if (config.issuer) {
                verifyOptions.issuer = config.issuer;
            }
            const secretValue = unwrapSecret(config.secret);
            const payload = jwt.verify(token, secretValue, verifyOptions);
            // 3. Custom validation
            if (config.customValidation && !config.customValidation(payload)) {
                return res.status(403).json({
                    error: 'forbidden',
                    message: 'Token validation failed',
                });
            }
            // 4. Attach to request
            req.auth = {
                authenticated: true,
                tokenInfo: {
                    active: true,
                    sub: payload.sub,
                    aud: typeof payload.aud === 'string' ? [payload.aud] : payload.aud,
                    iss: payload.iss,
                    exp: payload.exp,
                    iat: payload.iat,
                },
                scopes: payload.scopes || payload.scope?.split(' ') || [],
                clientId: payload.client_id || payload.sub,
                subject: payload.sub,
            };
            next();
        }
        catch (error) {
            const err = error;
            if (err.name === 'TokenExpiredError') {
                return res.status(401).json({
                    error: 'token_expired',
                    message: 'JWT has expired',
                });
            }
            if (err.name === 'JsonWebTokenError') {
                return res.status(401).json({
                    error: 'invalid_token',
                    message: 'Invalid JWT: ' + err.message,
                });
            }
            return res.status(500).json({
                error: 'server_error',
                message: 'Token validation failed',
            });
        }
    };
}
/**
 * Generate a JWT token
 *
 * @example
 * ```typescript
 * const token = generateJWT({
 *   secret: SecretValue.fromEnv('JWT_SECRET'),
 *   payload: {
 *     sub: 'user123',
 *     scopes: ['mcp:read', 'mcp:write'],
 *   },
 *   expiresIn: '1h',
 * });
 * ```
 */
export function generateJWT(options) {
    const signOptions = {
        algorithm: options.algorithm || 'HS256',
    };
    if (options.expiresIn !== undefined) {
        // Cast to jwt.SignOptions['expiresIn'] which accepts string | number
        signOptions.expiresIn = options.expiresIn;
    }
    if (options.audience) {
        signOptions.audience = options.audience;
    }
    if (options.issuer) {
        signOptions.issuer = options.issuer;
    }
    const secretValue = unwrapSecret(options.secret);
    return jwt.sign(options.payload, secretValue, signOptions);
}
/**
 * Verify a JWT token without middleware (helper)
 *
 * @param token - The JWT token to verify
 * @param config - JWT configuration including secret
 * @returns The decoded payload if valid, null otherwise
 */
export function verifyJWT(token, config) {
    try {
        const verifyOptions = {
            algorithms: [config.algorithm || 'HS256'],
        };
        if (config.audience) {
            verifyOptions.audience = config.audience;
        }
        if (config.issuer) {
            verifyOptions.issuer = config.issuer;
        }
        const secretValue = unwrapSecret(config.secret);
        const payload = jwt.verify(token, secretValue, verifyOptions);
        if (config.customValidation && !config.customValidation(payload)) {
            return null;
        }
        return payload;
    }
    catch {
        return null;
    }
}
/**
 * Decode JWT without verification (for debugging)
 */
export function decodeJWT(token) {
    try {
        return jwt.decode(token);
    }
    catch {
        return null;
    }
}
