// @nitrostack/core/dist/core/middleware/middleware.decorator.js
import 'reflect-metadata';
const MIDDLEWARE_KEY = 'nitrostack:middleware';
const IS_MIDDLEWARE_KEY = 'nitrostack:is_middleware';
/**
 * Marks a class as a middleware
 *
 * @example
 * ```typescript
 * @Middleware()
 * export class LoggingMiddleware implements MiddlewareInterface {
 *   async use(context: ExecutionContext, next: () => Promise<unknown>) {
 *     console.log('Before');
 *     const result = await next();
 *     console.log('After');
 *     return result;
 *   }
 * }
 * ```
 */
export function Middleware() {
    return (target) => {
        Reflect.defineMetadata(IS_MIDDLEWARE_KEY, true, target);
    };
}
/**
 * Apply middleware to a tool method
 *
 * @example
 * ```typescript
 * @Tool({ name: 'my_tool', ... })
 * @UseMiddleware(LoggingMiddleware, AuthMiddleware)
 * async myTool(input: Record<string, unknown>, context: ExecutionContext) { }
 * ```
 */
export function UseMiddleware(...middlewares) {
    return (target, propertyKey, descriptor) => {
        const existingMiddlewares = Reflect.getMetadata(MIDDLEWARE_KEY, target, propertyKey) || [];
        Reflect.defineMetadata(MIDDLEWARE_KEY, [...existingMiddlewares, ...middlewares], target, propertyKey);
    };
}
/**
 * Get middlewares for a method
 */
export function getMiddlewareMetadata(target, propertyKey) {
    return Reflect.getMetadata(MIDDLEWARE_KEY, target, propertyKey) || [];
}
/**
 * Check if a class is marked as middleware
 */
export function isMiddleware(target) {
    return Reflect.getMetadata(IS_MIDDLEWARE_KEY, target) === true;
}
