// @nitrostack/core/dist/core/pipes/pipe.decorator.js
import 'reflect-metadata';
const PIPE_KEY = 'nitrostack:pipe';
const IS_PIPE_KEY = 'nitrostack:is_pipe';
const PARAM_PIPES_KEY = 'nitrostack:param_pipes';
/**
 * Marks a class as a pipe
 *
 * @example
 * ```typescript
 * @Pipe()
 * export class ValidationPipe implements PipeInterface {
 *   transform(value: unknown, metadata: ArgumentMetadata) {
 *     // Validate and transform
 *     return value;
 *   }
 * }
 * ```
 */
export function Pipe() {
    return (target) => {
        Reflect.defineMetadata(IS_PIPE_KEY, true, target);
    };
}
/**
 * Apply pipes to tool input (entire input object)
 *
 * @example
 * ```typescript
 * @Tool({ name: 'create_user', ... })
 * @UsePipes(ValidationPipe, TransformPipe)
 * async createUser(input: unknown) { }
 * ```
 */
export function UsePipes(...pipes) {
    return (target, propertyKey, descriptor) => {
        const existingPipes = Reflect.getMetadata(PIPE_KEY, target, propertyKey) || [];
        Reflect.defineMetadata(PIPE_KEY, [...existingPipes, ...pipes], target, propertyKey);
    };
}
/**
 * Parameter decorator to apply pipes to specific parameters
 *
 * @example
 * ```typescript
 * @Tool({ name: 'create_user', ... })
 * async createUser(@Body(ValidationPipe) input: CreateUserDto) { }
 * ```
 */
export function Body(...pipes) {
    return (target, propertyKey, parameterIndex) => {
        if (!propertyKey)
            return;
        // Clone to avoid mutating a parent class's param-pipe object when subclassing.
        const existingParams = { ...(Reflect.getMetadata(PARAM_PIPES_KEY, target, propertyKey) || {}) };
        existingParams[parameterIndex] = {
            type: 'body',
            pipes,
        };
        Reflect.defineMetadata(PARAM_PIPES_KEY, existingParams, target, propertyKey);
    };
}
/**
 * Shorthand for validation pipe
 */
export function Validated() {
    return Body(); // Can be enhanced with default validation pipe
}
/**
 * Get pipes for a method
 */
export function getPipeMetadata(target, propertyKey) {
    return Reflect.getMetadata(PIPE_KEY, target, propertyKey) || [];
}
/**
 * Get parameter pipes for a method
 */
export function getParamPipesMetadata(target, propertyKey) {
    return Reflect.getMetadata(PARAM_PIPES_KEY, target, propertyKey) || {};
}
/**
 * Check if a class is marked as a pipe
 */
export function isPipe(target) {
    return Reflect.getMetadata(IS_PIPE_KEY, target) === true;
}
