// @nitrostack/core/dist/core/transports/discovery-http-server.js
import http from 'http';
import net from 'net';
/**
 * Check if a port is available
 */
async function isPortAvailable(port) {
    return new Promise((resolve) => {
        const server = net.createServer();
        server.once('error', () => resolve(false));
        server.once('listening', () => {
            server.close(() => resolve(true));
        });
        server.listen(port);
    });
}
/**
 * Find an available port starting from the preferred port
 */
async function findAvailablePort(startPort, maxRetries) {
    for (let i = 0; i < maxRetries; i++) {
        const port = startPort + i;
        if (await isPortAvailable(port)) {
            return port;
        }
    }
    throw new Error(`No available port found in range ${startPort}-${startPort + maxRetries - 1}`);
}
export class DiscoveryHttpServer {
    logger;
    server;
    handlers = new Map();
    _actualPort;
    options;
    constructor(portOrOptions, logger) {
        this.logger = logger;
        // Support both old (port number) and new (options object) API
        if (typeof portOrOptions === 'number') {
            this.options = {
                port: portOrOptions,
                autoRetry: false,
                maxRetries: 10,
            };
        }
        else {
            this.options = {
                port: portOrOptions.port ?? 3005,
                autoRetry: portOrOptions.autoRetry ?? true,
                maxRetries: portOrOptions.maxRetries ?? 10,
            };
        }
        this._actualPort = this.options.port;
        this.server = http.createServer(this.handleRequest.bind(this));
    }
    handleRequest(req, res) {
        this.logger.info(`DiscoveryHttpServer: Received request for ${req.url}`);
        const handler = this.handlers.get(req.url || '');
        if (handler) {
            handler(req, res);
        }
        else {
            this.logger.warn(`DiscoveryHttpServer: No handler found for ${req.url}`);
            res.writeHead(404);
            res.end();
        }
    }
    on(path, handler) {
        this.logger.info(`DiscoveryHttpServer: Registering handler for ${path}`);
        this.handlers.set(path, handler);
    }
    /**
     * Get the actual port the server is running on
     * (may differ from configured port if auto-retry found an available port)
     */
    getPort() {
        return this._actualPort;
    }
    async start() {
        // Find available port if auto-retry is enabled
        if (this.options.autoRetry) {
            try {
                this._actualPort = await findAvailablePort(this.options.port, this.options.maxRetries);
                if (this._actualPort !== this.options.port) {
                    this.logger.info(`DiscoveryHttpServer: Port ${this.options.port} in use, using ${this._actualPort}`);
                }
            }
            catch (error) {
                this.logger.error(`DiscoveryHttpServer: Failed to find available port`, error);
                throw error;
            }
        }
        return new Promise((resolve, reject) => {
            this.server.listen(this._actualPort, () => {
                this.logger.info(`🚀 OAuth discovery server running at http://localhost:${this._actualPort}`);
                resolve();
            });
            this.server.on('error', (err) => {
                if (err.code === 'EADDRINUSE' && !this.options.autoRetry) {
                    this.logger.error(`OAuth discovery server error: Port ${this._actualPort} is already in use. ` +
                        `Set OAUTH_DISCOVERY_PORT to a different port or enable auto-retry.`);
                }
                else {
                    this.logger.error('OAuth discovery server error', err);
                }
                reject(err);
            });
        });
    }
    stop() {
        return new Promise((resolve) => {
            if (this.server.listening) {
                this.server.close(() => {
                    this.logger.info('OAuth discovery server stopped');
                    resolve();
                });
            }
            else {
                resolve();
            }
        });
    }
}
