// @nitrostack/core/dist/core/widgets/widget-registry.js
/**
 * Widget Registry
 *
 * Manages widget metadata and example resolution
 */
import { readFileSync, existsSync } from 'fs';
/**
 * Widget Registry
 * Loads and manages widget metadata from manifest
 */
export class WidgetRegistry {
    widgets = new Map();
    loaded = false;
    /**
     * Load widget manifest from file
     */
    loadManifest(manifestPath) {
        if (!existsSync(manifestPath)) {
            // Use stderr to avoid corrupting MCP stdio protocol
            console.error(`[Widget Registry] Manifest not found at ${manifestPath}, skipping widget examples`);
            return;
        }
        try {
            const manifestContent = readFileSync(manifestPath, 'utf-8');
            const manifest = JSON.parse(manifestContent);
            for (const widget of manifest.widgets) {
                this.widgets.set(widget.uri, widget);
            }
            this.loaded = true;
            // Use stderr to avoid corrupting MCP stdio protocol
            console.error(`[Widget Registry] Loaded ${this.widgets.size} widget(s) from manifest`);
        }
        catch (error) {
            console.error(`[Widget Registry] Failed to load widget manifest:`, error instanceof Error ? error.message : String(error));
        }
    }
    /**
     * Get all widgets
     */
    getAllWidgets() {
        return Array.from(this.widgets.values());
    }
    /**
     * Get widget by URI
     */
    getWidget(uri) {
        return this.widgets.get(uri);
    }
    /**
     * Get widget example data
     * Returns the first example's data, or undefined if no examples
     */
    getWidgetExampleData(uri) {
        const widget = this.widgets.get(uri);
        return widget?.examples[0]?.data;
    }
    /**
     * Check if manifest is loaded
     */
    isLoaded() {
        return this.loaded;
    }
}
// Global registry instance
let registryInstance = null;
/**
 * Get the global widget registry instance
 */
export function getWidgetRegistry() {
    if (!registryInstance) {
        registryInstance = new WidgetRegistry();
    }
    return registryInstance;
}
