// @nitrostack/core/dist/core/events/event-emitter.js
import 'reflect-metadata';
/**
 * Event Emitter
 *
 * Simple event bus for decoupled communication between modules
 */
export class EventEmitter {
    static instance;
    listeners = new Map();
    constructor() { }
    static getInstance() {
        if (!EventEmitter.instance) {
            EventEmitter.instance = new EventEmitter();
        }
        return EventEmitter.instance;
    }
    /**
     * Register an event listener
     */
    on(event, handler) {
        if (!this.listeners.has(event)) {
            this.listeners.set(event, []);
        }
        this.listeners.get(event).push(handler);
    }
    /**
     * Register a one-time event listener
     */
    once(event, handler) {
        const wrappedHandler = async (...args) => {
            await handler(...args);
            this.off(event, wrappedHandler);
        };
        this.on(event, wrappedHandler);
    }
    /**
     * Remove an event listener
     */
    off(event, handler) {
        const handlers = this.listeners.get(event);
        if (!handlers)
            return;
        const index = handlers.indexOf(handler);
        if (index > -1) {
            handlers.splice(index, 1);
        }
        if (handlers.length === 0) {
            this.listeners.delete(event);
        }
    }
    /**
     * Emit an event
     */
    async emit(event, ...args) {
        const handlers = this.listeners.get(event);
        if (!handlers || handlers.length === 0)
            return;
        // Execute all handlers in parallel
        await Promise.all(handlers.map(handler => handler(...args)));
    }
    /**
     * Emit an event synchronously
     */
    emitSync(event, ...args) {
        const handlers = this.listeners.get(event);
        if (!handlers || handlers.length === 0)
            return;
        handlers.forEach(handler => handler(...args));
    }
    /**
     * Remove all listeners for an event
     */
    removeAllListeners(event) {
        if (event) {
            this.listeners.delete(event);
        }
        else {
            this.listeners.clear();
        }
    }
    /**
     * Get listener count for an event
     */
    listenerCount(event) {
        return this.listeners.get(event)?.length || 0;
    }
    /**
     * Get all event names
     */
    eventNames() {
        return Array.from(this.listeners.keys());
    }
}
