// @nitrostack/core/dist/testing/index.js
import 'reflect-metadata';
import { DIContainer } from '../core/di/container.js';
/**
 * Mock logger for testing
 */
export class MockLogger {
    logs = [];
    debug(message, meta) {
        this.logs.push({ level: 'debug', message, meta });
    }
    info(message, meta) {
        this.logs.push({ level: 'info', message, meta });
    }
    warn(message, meta) {
        this.logs.push({ level: 'warn', message, meta });
    }
    error(message, meta) {
        this.logs.push({ level: 'error', message, meta });
    }
    clear() {
        this.logs = [];
    }
    hasLog(level, message) {
        return this.logs.some(log => log.level === level && log.message.includes(message));
    }
}
/**
 * Create a mock execution context for testing
 */
export function createMockContext(overrides) {
    const mockLogger = new MockLogger();
    return {
        requestId: 'test-request-id',
        logger: mockLogger,
        metadata: {},
        auth: undefined,
        ...overrides,
    };
}
/**
 * Testing module builder
 */
export class TestingModule {
    container;
    providers = new Map();
    mocks = new Map();
    constructor() {
        this.container = DIContainer.getInstance();
    }
    /**
     * Create a testing module
     */
    static create() {
        return new TestingModule();
    }
    /**
     * Add a provider to the module
     */
    addProvider(token, provider) {
        this.providers.set(token, (provider || token));
        return this;
    }
    /**
     * Add a mock to the module
     */
    addMock(token, mock) {
        this.mocks.set(token, mock);
        return this;
    }
    /**
     * Compile the testing module
     */
    compile() {
        // Clear existing container
        this.container.clear();
        // Register mocks first (they take precedence)
        for (const [token, mock] of this.mocks.entries()) {
            this.container.registerValue(token, mock);
        }
        // Register providers
        for (const [token, provider] of this.providers.entries()) {
            if (!this.mocks.has(token)) {
                this.container.register(token, provider);
            }
        }
        return new CompiledTestingModule(this.container);
    }
}
/**
 * Compiled testing module
 */
export class CompiledTestingModule {
    container;
    constructor(container) {
        this.container = container;
    }
    /**
     * Get an instance from the container
     */
    get(token) {
        return this.container.resolve(token);
    }
    /**
     * Cleanup the module
     */
    cleanup() {
        this.container.clear();
    }
}
/**
 * Create a mock function (compatible with Jest)
 */
export function createMockFn() {
    // If jest is available, use it
    if (typeof jest !== 'undefined' && jest?.fn) {
        return jest.fn();
    }
    // Otherwise, create a simple mock
    const fn = ((...args) => {
        fn.calls.push(args);
        return fn.mockReturnValue;
    });
    fn.calls = [];
    fn.mockReturnValue = undefined;
    fn.mockResolvedValue = (value) => {
        fn.mockReturnValue = Promise.resolve(value);
        return fn;
    };
    fn.mockRejectedValue = (error) => {
        fn.mockReturnValue = Promise.reject(error);
        return fn;
    };
    return fn;
}
/**
 * Wait for async operations to complete
 */
export async function flushPromises() {
    return new Promise(resolve => setImmediate(resolve));
}
/**
 * Spy on a method
 */
export function spyOn(obj, method) {
    if (typeof jest !== 'undefined' && jest?.spyOn) {
        return jest.spyOn(obj, method);
    }
    // Simple spy implementation
    const original = obj[method];
    const spy = ((...args) => {
        spy.calls.push(args);
        return original.apply(obj, args);
    });
    spy.calls = [];
    spy.restore = () => {
        obj[method] = original;
    };
    obj[method] = spy;
    return spy;
}
